{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from PIL import ImageTk, Image  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from tkinter import *\n",
    "from PIL import Image, ImageTk\n",
    "import pymongo\n",
    "from tkinter import messagebox\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib.backends.backend_tkagg import ( FigureCanvasTkAgg, NavigationToolbar2Tk) \n",
    "from matplotlib.figure import Figure\n",
    "from wordcloud import WordCloud\n",
    "import pymongo\n",
    "from pymongo import MongoClient\n",
    "from tkinter.ttk import Combobox\n",
    "from tkinter import Toplevel, Text\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Toxins Data Insertion and Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mechanism = pd.read_csv('Relations_table1.csv')\n",
    "target = pd.read_csv('AllToxinTarget.csv' , low_memory=False)\n",
    "edited = pd.read_csv('Mechanism.csv' ,low_memory=False)\n",
    "structures = pd.read_csv('ToxinStructures.csv')\n",
    "data =pd.read_csv('AllToxinData.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "client = MongoClient('mongodb://localhost:27017/')\n",
    "db = client['toxins']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x1e7e72bb460>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_toxin_target_collection = db['AllToxinTarget']\n",
    "all_toxin_target=target.to_dict(orient='records')\n",
    "#all_toxin_target\n",
    "all_toxin_target_collection.insert_many(all_toxin_target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "edited_json_data=edited.to_dict(orient='records')\n",
    "structure_json_data=structures.to_dict(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x1e7e72bb550>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_toxin_data_collection = db['AllToxinData']\n",
    "all_toxin_data=data.to_dict(orient='records')\n",
    "#all_toxin_data\n",
    "all_toxin_data_collection.insert_many(all_toxin_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x1e7e72a1be0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_toxin_mec_collection = db['Mechanism']\n",
    "all_toxin_mec=mechanism.to_dict(orient='records')\n",
    "#all_toxin_data\n",
    "all_toxin_mec_collection.insert_many(all_toxin_mec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "for entry in all_toxin_data:\n",
    "    del (entry['mechanism_of_toxicity'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.command_cursor.CommandCursor at 0x1e7e747bd00>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_toxin_mec_collection.aggregate( [{\"$match\":{\"mechanismID\":{ \"$ne\": \"_\" }}},\n",
    "    { \"$group\" : { \"_id\" : \"$Toxin T3DB ID\", \"target-mechanism\": { \"$push\": {\"targetID\":\"$Target UniProt ID\",\"mechanismID\":\"$mechanismID\" } } } },\n",
    "    { \"$out\" : \"authors\" }\n",
    "] )\n",
    "\n",
    "all_toxin_mec_collection.aggregate( [{\"$match\":{\"mechanismID\":{ \"$ne\": \"_\" }}},\n",
    "    { \"$group\" : { \"_id\" : \"$Target UniProt ID\", \"toxin-mechanism\": { \"$push\": {\"toxinID\":\"$Toxin T3DB ID\",\"mechanismID\":\"$mechanismID\" } } } },\n",
    "    { \"$out\" : \"authors1\" }\n",
    "] )\n",
    "\n",
    "all_toxin_mec_collection.aggregate( [{\"$match\":{\"mechanismID\":{ \"$ne\": \"_\" }}},\n",
    "    { \"$group\" : { \"_id\" : \"$mechanismID\", \"toxin-target\": { \"$push\": {\"toxinID\":\"$Toxin T3DB ID\",\"targetID\":\"$targetID\" } } } },\n",
    "    { \"$out\" : \"authors2\" }\n",
    "] )\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "toxinCollectionData = all_toxin_data_collection.find()\n",
    "targetCollectionData = all_toxin_target_collection.find()\n",
    "# Exception here we clear mechanism to be ready for the relation\n",
    "all_toxin_mec_collection.drop()\n",
    "\n",
    "all_toxin_mec_collection.insert_many(edited_json_data)\n",
    "\n",
    "mechanismCollectionData = all_toxin_mec_collection.find()\n",
    "\n",
    "\n",
    "# recieving the outputs here\n",
    "\n",
    "\n",
    "authorsCollection = db['authors']\n",
    "\n",
    "authorsCollection1 = db['authors1']\n",
    "\n",
    "authorsCollection2 = db['authors2']\n",
    "\n",
    "\n",
    "for entry in toxinCollectionData:\n",
    "    \n",
    "    x = authorsCollection.find({\"_id\":entry['_id']}).distinct(\"target-mechanism\")\n",
    "    \n",
    "    \n",
    "    all_toxin_data_collection.update_one({\"_id\":entry['_id']},{\"$set\":{\"target-mechanism\":x}})\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "for entry in targetCollectionData:\n",
    "    x = authorsCollection1.find({\"_id\":entry['_id']}).distinct(\"toxin-mechanism\")\n",
    "    \n",
    "    all_toxin_target_collection.update_one({\"_id\":entry['_id']},{\"$set\":{\"toxin-mechanism\":x}})\n",
    "    \n",
    "#     print(x)\n",
    "\n",
    "for entry in structure_json_data:\n",
    "    all_toxin_data_collection.update_one({\"_id\":entry['DATABASE_ID']},{\"$set\":{\"toxin_structure\":entry}})\n",
    "\n",
    "\n",
    "\n",
    "for entry in mechanismCollectionData:\n",
    "    x = authorsCollection2.find({\"_id\":entry['_id']}).distinct(\"toxin-target\")\n",
    "    \n",
    "    all_toxin_mec_collection.update_one({\"_id\":entry['_id']},{\"$set\":{\"toxin-target\":x}})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "from pymongo import MongoClient\n",
    "\n",
    "# Global variables for entry fields and result text widget\n",
    "old_synonyms_entry = None\n",
    "new_synonyms_entry = None\n",
    "result_text = None\n",
    "\n",
    "# Function to update synonyms in the toxin target collection\n",
    "def update_synonyms():\n",
    "    old_synonyms = old_synonyms_entry.get()\n",
    "    new_synonyms = new_synonyms_entry.get()\n",
    "\n",
    "    myquery = {\"Synonyms\": old_synonyms}\n",
    "    newvalues = {\"$set\": {\"Synonyms\": new_synonyms}}\n",
    "\n",
    "    all_toxin_target_collection.update_one(myquery, newvalues)\n",
    "\n",
    "    documents = all_toxin_target_collection.find({\"Synonyms\": new_synonyms}).limit(3)\n",
    "    result = list(documents)\n",
    "    df = pd.DataFrame(result)\n",
    "    result_text.config(state=NORMAL)\n",
    "    result_text.delete(\"1.0\", END)\n",
    "    result_text.insert(END, df.to_string(index=False))\n",
    "    result_text.config(state=DISABLED)\n",
    "\n",
    "# Function to open the update screen\n",
    "def open_update_screen():\n",
    "    global old_synonyms_entry, new_synonyms_entry, result_text  # Make the variables global\n",
    "\n",
    "    update_screen = Toplevel(root)\n",
    "    update_screen.title(\"Update Synonyms\")\n",
    "    update_screen.configure(bg='light blue')  # Set background color to light blue\n",
    "    update_screen.geometry(\"400x200\")  # Increase screen size\n",
    "\n",
    "    # Create labels\n",
    "    old_synonyms_label = Label(update_screen, text=\"Old Synonyms:\", bg='light blue')\n",
    "    old_synonyms_label.pack()\n",
    "    new_synonyms_label = Label(update_screen, text=\"New Synonyms:\", bg='light blue')\n",
    "    new_synonyms_label.pack()\n",
    "\n",
    "    # Create entry fields\n",
    "    old_synonyms_entry = Entry(update_screen)\n",
    "    old_synonyms_entry.pack()\n",
    "    new_synonyms_entry = Entry(update_screen)\n",
    "    new_synonyms_entry.pack()\n",
    "\n",
    "    # Create update button\n",
    "    update_button = Button(update_screen, text=\"Update\", command=update_synonyms)\n",
    "    update_button.pack()\n",
    "\n",
    "    # Create result text widget\n",
    "    result_text = tk.Text(update_screen, height=10, width=40, state=tk.DISABLED)\n",
    "    result_text.pack()\n",
    "\n",
    "# Connect to MongoDB\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Insertion "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insertion_screen():\n",
    "    global insertion_database\n",
    "    insertion_database = Toplevel(root)\n",
    "    insertion_database.title(\"Toxins\")\n",
    "    insertion_database.geometry(\"400x200\")\n",
    "    insertion_database.configure(bg=\"lavender\")\n",
    "\n",
    "    # Create the main text label\n",
    "    main_label = Label(insertion_database, text=\"Data Insertion Window\", font=(\"Arial\", 14, \"bold\"), bg=\"lavender\")\n",
    "    main_label.pack(pady=20)\n",
    "\n",
    "    # Create the buttons frame\n",
    "    buttons_frame = Frame(insertion_database, bg=\"lavender\")\n",
    "    buttons_frame.pack()\n",
    "\n",
    "    # Create the buttons\n",
    "    button_all_data = Button(buttons_frame, text=\"All Toxins Data\", width=20, command=dataentry)\n",
    "    button_all_data.pack(side=LEFT, padx=10)\n",
    "\n",
    "    button_all_target = Button(buttons_frame, text=\"All Toxins Target\", width=20, command=insert_toxinTarget)\n",
    "    button_all_target.pack(side=LEFT, padx=10)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def msgbox(msg,title):\n",
    "    result=messagebox.askokcancel(title=title,message=msg)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#functions\n",
    "\n",
    "def save():\n",
    "    r=msgbox(\"Save record?\",\"record\")\n",
    "    if r==True:\n",
    "        newid=mytable.count_documents({})\n",
    "        if newid!=0:\n",
    "            \n",
    "            newid=mytable.find_one(sort=[(\"_id\",-1)])[\"id\"]\n",
    "        id=newid+1   \n",
    "        cid.set(id)\n",
    "        mydict={\"_id\":int(id.get()),\"Name\":Name.get(),\"studemail\":uniprot_name.get(),\"student course\":ccourse.get()}\n",
    "        x=mytable.insert_one(mydict)\n",
    "#delete function\n",
    "\n",
    "def delete():\n",
    "    r=msgbox(\"delete record?\",\"record\")\n",
    "    if r==True:\n",
    "        myquery={\"_id\":int(delId.get())}\n",
    "        mytable.delete_one(myquery)\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert Many functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_many(mylist):\n",
    "    # Connect to the MongoDB server and select the database and collection\n",
    "    client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "    db = client[\"toxins\"]\n",
    "    all_toxin_data_collection = db[\"AllToxinData\"]\n",
    "\n",
    "    # Insert the list of documents into the collection\n",
    "    all_toxin_data_collection.insert_many(mylist)\n",
    "\n",
    "def open_insert_window():\n",
    "    def insert_documents():\n",
    "        documents = []\n",
    "        for entry in entries:\n",
    "            document = {\"_id\": entry[\"_id\"].get(), \"chemical_formula\": entry[\"chemical_formula\"].get()}\n",
    "            documents.append(document)\n",
    "\n",
    "        insert_many(documents)\n",
    "        insert_button.config(state=tk.DISABLED)\n",
    "\n",
    "    # Create insert window\n",
    "    insert_window = Toplevel(root)\n",
    "    insert_window.title(\"Insert Toxin Data\")\n",
    "    insert_window.configure(bg='light blue')\n",
    "    insert_window.geometry(\"400x250\")\n",
    "\n",
    "    entries = []\n",
    "\n",
    "    # Create entry fields\n",
    "    for i in range(3):\n",
    "        entry_frame = Frame(insert_window, bg='light blue')\n",
    "        entry_frame.pack(pady=10)\n",
    "\n",
    "        _id_label = Label(entry_frame, text=\"Enter _id:\")\n",
    "        _id_label.pack(side=tk.LEFT)\n",
    "        _id_entry = Entry(entry_frame)\n",
    "        _id_entry.pack(side=tk.LEFT)\n",
    "\n",
    "        chemical_formula_label = Label(entry_frame, text=\"Enter chemical formula:\")\n",
    "        chemical_formula_label.pack(side=tk.LEFT)\n",
    "        chemical_formula_entry = Entry(entry_frame)\n",
    "        chemical_formula_entry.pack(side=tk.LEFT)\n",
    "\n",
    "        entry = {\"_id\": _id_entry, \"chemical_formula\": chemical_formula_entry}\n",
    "        entries.append(entry)\n",
    "\n",
    "    # Create insert button\n",
    "    insert_button = tk.Button(insert_window, text=\"Insert\", command=insert_documents)\n",
    "    insert_button.pack(pady=20)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tkinter import Tk, Label, Entry, Button, Toplevel\n",
    "\n",
    "\n",
    "############all toxins data collection\n",
    "def update_common_name():\n",
    "    updateWindow = Toplevel(root)\n",
    "    updateWindow.title(\"Update Common Name\")\n",
    "    updateWindow.geometry(\"400x200\")\n",
    "\n",
    "    old_common_name_label = Label(updateWindow, text=\"Old Common Name:\", bg=\"lavender\", fg=\"grey\")\n",
    "    old_common_name_label.pack(pady=10)\n",
    "    old_common_name_entry = Entry(updateWindow)\n",
    "    old_common_name_entry.pack(pady=10)\n",
    "\n",
    "    new_common_name_label = Label(updateWindow, text=\"New Common Name:\", bg=\"lavender\", fg=\"grey\")\n",
    "    new_common_name_label.pack(pady=10)\n",
    "    new_common_name_entry = Entry(updateWindow)\n",
    "    new_common_name_entry.pack(pady=10)\n",
    "\n",
    "    def update():\n",
    "        old_common_name = old_common_name_entry.get()\n",
    "        new_common_name = new_common_name_entry.get()\n",
    "\n",
    "        query = {\"common_name\": old_common_name}\n",
    "        new_values = {\"$set\": {\"common_name\": new_common_name}}\n",
    "\n",
    "        try:\n",
    "            result = all_toxin_data_collection.update_many(query, new_values)\n",
    "            if result.modified_count > 0:\n",
    "                messagebox.showinfo(\"Update Successful\", \"Common name updated successfully.\")\n",
    "            else:\n",
    "                messagebox.showwarning(\"No Records Updated\", \"No matching records found for update.\")\n",
    "        except Exception as e:\n",
    "            messagebox.showerror(\"Update Error\", f\"An error occurred during update: {str(e)}\")\n",
    "\n",
    "        updateWindow.destroy()\n",
    "\n",
    "    update_button = Button(updateWindow, text=\"Update\", command=update)\n",
    "    update_button.pack(pady=10)\n",
    "\n",
    "\n",
    "\n",
    "def display_max_weight():\n",
    "    # Call the get_max_weight function to retrieve the maximum weight document\n",
    "    max_weight_doc = get_max_weight(all_toxin_data_collection)\n",
    "    \n",
    "    if max_weight_doc:\n",
    "        # Create a formatted message to display the result\n",
    "        message = f\"Max Weight Document:\\n{max_weight_doc}\"\n",
    "    else:\n",
    "        message = \"No documents found.\"\n",
    "    \n",
    "    # Open a message box to display the result\n",
    "    messagebox.showinfo(\"Maximum Weight\", message)\n",
    "\n",
    "def open_update_window():\n",
    "    update_window = Toplevel(root)\n",
    "    update_window.title(\"Update Window\")\n",
    "    update_window.geometry(\"600x400\")\n",
    "\n",
    "    # Load the background image\n",
    "    bg_image = ImageTk.PhotoImage(Image.open(\"find.PNG\"))\n",
    "\n",
    "    # Create a canvas with the background image\n",
    "    canvas = Canvas(update_window, width=600, height=400)\n",
    "    canvas.pack(fill=\"both\", expand=True)\n",
    "    canvas.create_image(0, 0, image=bg_image, anchor=\"nw\")\n",
    "\n",
    "    # Create buttons in the center of the window\n",
    "    button1 = Button(update_window, text=\"update common_name\",width=20,bg=\"#63a7c1\",command=update_common_name)\n",
    "    button1.place(relx=0.5, rely=0.3, anchor=CENTER)\n",
    "\n",
    "    button2 = Button(update_window, text=\"update health effect\", bg=\"#63a7c1\",command=open_update_window_health_effect)\n",
    "    button2.place(relx=0.5, rely=0.4, anchor=CENTER)\n",
    "\n",
    "    button3 = Button(update_window, text=\"update date\", bg=\"#63a7c1\",command=update_toxin_data_gui)\n",
    "    button3.place(relx=0.5, rely=0.5, anchor=CENTER)\n",
    "\n",
    "#     button4 = Button(update_window, text=\"update target data\", bg=\"#63a7c1\",command=open_update_target_data)\n",
    "#     button4.place(relx=0.5, rely=0.6, anchor=CENTER)\n",
    "\n",
    "    button5 = Button(update_window, text=\"close\", bg=\"#63a7c1\",command=update_window.destroy)\n",
    "    button5.place(relx=0.5, rely=0.7, anchor=CENTER)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def dataentry():\n",
    "    dataInsertionwindow = Toplevel(root, bg=\"lavender\")\n",
    "    dataInsertionwindow.title(\"Data Insertion Window\")\n",
    "    dataInsertionwindow.geometry(\"800x700\")\n",
    "\n",
    "    label = Label(dataInsertionwindow, bg=\"lavender\", text=\"Data Insertion Form\", width=30, height=3, anchor=\"center\")\n",
    "    label.grid(row=0, column=0, columnspan=2, pady=10)\n",
    "\n",
    "    # Create labels and entry fields\n",
    "    id_label = Label(dataInsertionwindow, text=\"ID:\", bg=\"lavender\", fg=\"grey\")\n",
    "    id_label.grid(row=1, column=0, padx=10, pady=10, sticky=\"e\")\n",
    "    id_entry = Entry(dataInsertionwindow)\n",
    "    id_entry.grid(row=1, column=1, padx=10, pady=10)\n",
    "\n",
    "    common_name_label = Label(dataInsertionwindow, text=\"Common Name:\", bg=\"lavender\", fg=\"grey\")\n",
    "    common_name_label.grid(row=2, column=0, padx=10, pady=10, sticky=\"e\")\n",
    "    common_name_entry = Entry(dataInsertionwindow)\n",
    "    common_name_entry.grid(row=2, column=1, padx=10, pady=10)\n",
    "\n",
    "    description_new_label = Label(dataInsertionwindow, text=\"Description:\", bg=\"lavender\", fg=\"grey\")\n",
    "    description_new_label.grid(row=3, column=0, padx=10, pady=10, sticky=\"e\")\n",
    "    description_Entry = Entry(dataInsertionwindow)\n",
    "    description_Entry.grid(row=3, column=1, padx=10, pady=10)\n",
    "\n",
    "    chemical_formula_new_label = Label(dataInsertionwindow, text=\"Chemical Formula:\", bg=\"lavender\", fg=\"grey\")\n",
    "    chemical_formula_new_label.grid(row=4, column=0, padx=10, pady=10, sticky=\"e\")\n",
    "    chemical_formula_Entry = Entry(dataInsertionwindow)\n",
    "    chemical_formula_Entry.grid(row=4, column=1, padx=10, pady=10)\n",
    "\n",
    "    delete_id_label = Label(dataInsertionwindow, text=\"Delete ID:\", bg=\"lavender\", fg=\"grey\")\n",
    "    delete_id_label.grid(row=5, column=0, padx=10, pady=10, sticky=\"e\")\n",
    "    delete_id_entry = Entry(dataInsertionwindow)\n",
    "    delete_id_entry.grid(row=5, column=1, padx=10, pady=10)\n",
    "\n",
    "    weight_label = Label(dataInsertionwindow, text=\"Weight:\", bg=\"lavender\", fg=\"grey\")\n",
    "    weight_label.grid(row=6, column=0, padx=10, pady=10, sticky=\"e\")\n",
    "    weight_entry = Entry(dataInsertionwindow)\n",
    "    weight_entry.grid(row=6, column=1, padx=10, pady=10)\n",
    "\n",
    "    def clear_entries():\n",
    "        id_entry.delete(0, 'end')\n",
    "        common_name_entry.delete(0, 'end')\n",
    "        description_Entry.delete(0, 'end')\n",
    "        chemical_formula_Entry.delete(0, 'end')\n",
    "        delete_id_entry.delete(0, 'end')\n",
    "        weight_entry.delete(0, 'end')\n",
    "\n",
    "    def delete():\n",
    "        r = messagebox.askyesno(\"Delete Record\", \"Are you sure you want to delete the record?\")\n",
    "        if r:\n",
    "            myquery = {\"_id\": delete_id_entry.get()}\n",
    "            all_toxin_data_collection.delete_one(myquery)\n",
    "            clear_entries()\n",
    "\n",
    "    def insert_data():\n",
    "        document = {\n",
    "            \"_id\": id_entry.get(),\n",
    "            \"common_name\": common_name_entry.get(),\n",
    "            \"description\": description_Entry.get(),\n",
    "            \"chemical_formula\": chemical_formula_Entry.get(),\n",
    "            \"weight\": weight_entry.get()\n",
    "        }\n",
    "        all_toxin_data_collection.insert_one(document)\n",
    "        clear_entries()\n",
    "\n",
    "\n",
    "    insert_button = Button(dataInsertionwindow, width=25, height=2, bg=\"grey\", fg=\"white\", text=\"Save\", command=insert_data)\n",
    "    insert_button.grid(row=7, column=1, padx=10, pady=10, sticky=\"e\")\n",
    "\n",
    "    delete_button = Button(dataInsertionwindow, width=25, height=2, bg=\"grey\", fg=\"white\", text=\"Delete\", command=delete)\n",
    "    delete_button.grid(row=8, column=1, padx=10, pady=10, sticky=\"e\")\n",
    "    \n",
    "\n",
    "\n",
    "    #find_btn = Button(dataInsertionwindow, width=25, height=2, bg=\"grey\", fg=\"white\", text=\"Find Queries\", command=open_find_queries_window)\n",
    "    #find_btn.grid(row=12, column=1, padx=10, pady=10, sticky=\"e\")\n",
    "\n",
    "    #update_button = Button(dataInsertionwindow, width=25, height=2, bg=\"grey\", fg=\"white\", text=\"Update\", command=open_update_window)\n",
    "    #update_button.grid(row=14, column=1, padx=10, pady=10, sticky=\"e\")\n",
    "    insert_many_button = Button(dataInsertionwindow, width=25, height=2, bg=\"grey\", fg=\"white\", text=\"InsertMany \", command=open_insert_window)\n",
    "    insert_many_button.grid(row=16, column=1, padx=10, pady=10, sticky=\"e\")\n",
    "    close_button = Button(dataInsertionwindow, width=25, height=2, bg=\"grey\", fg=\"white\", text=\"Close\", command=dataInsertionwindow.destroy)\n",
    "    close_button.grid(row=18, column=1, padx=10, pady=10, sticky=\"e\")\n",
    "    #max_btn = Button(dataInsertionwindow, width=15, height=2, bg=\"grey\", fg=\"white\", text=\"Find Max Weight\", command=display_max_weight)\n",
    "    #max_btn.grid(row=11, column=1, padx=10, pady=10, sticky=\"e\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find query based on indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def indexing_find():\n",
    "    # Create the indexing_find window\n",
    "    indexing_window = Toplevel(root)\n",
    "    indexing_window.title(\"Indexing Find\")\n",
    "    indexing_window.geometry(\"800x600\")\n",
    "    indexing_window.configure(bg=\"lavender\")\n",
    "\n",
    "    # Create labels and radio buttons for sorting order\n",
    "    sort_label = Label(indexing_window, text=\"Sort Order:\", font=(\"Arial\", 12), bg=\"lavender\")\n",
    "    sort_label.pack()\n",
    "\n",
    "    sort_var = StringVar()\n",
    "    sort_var.set(\"Ascending\")\n",
    "\n",
    "    ascending_radio = Radiobutton(indexing_window, text=\"Ascending\", variable=sort_var, value=\"Ascending\", bg=\"lavender\")\n",
    "    ascending_radio.pack()\n",
    "\n",
    "    descending_radio = Radiobutton(indexing_window, text=\"Descending\", variable=sort_var, value=\"Descending\", bg=\"lavender\")\n",
    "    descending_radio.pack()\n",
    "\n",
    "    def show_result():\n",
    "        sort_order = 1 if sort_var.get() == \"Ascending\" else -1\n",
    "\n",
    "        # Create index on the \"Name\" field\n",
    "        all_toxin_target_collection.create_index([(\"Name\", sort_order)])\n",
    "\n",
    "        # Perform the query with the hint for index usage and limit the results to 10\n",
    "        result = all_toxin_target_collection.find().hint(\"Name_1\" if sort_order == 1 else \"Name_-1\").limit(10)\n",
    "        result = list(result)\n",
    "        df = pd.DataFrame(result)\n",
    "\n",
    "        # Display the DataFrame result\n",
    "        result_text.delete(\"1.0\", END)  # Clear previous result\n",
    "        result_text.insert(END, df)\n",
    "\n",
    "    # Create a button to show the result\n",
    "    show_button = Button(indexing_window, text=\"Show Result\", command=show_result, font=(\"Arial\", 12))\n",
    "    show_button.pack()\n",
    "\n",
    "    # Create a text widget to display the result\n",
    "    result_text = Text(indexing_window, height=20, width=100)\n",
    "    result_text.pack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### find query based in toxin state\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_dataframe():\n",
    "    # Get the selected value from the dropdown\n",
    "    query = dropdown.get()\n",
    "\n",
    "    pipeline = [\n",
    "        {\"$match\": {\"state\": query}},\n",
    "        {\"$project\": {\"_id\": 0, \"cas\": 1, \"common_name\": 1}},\n",
    "        {\"$limit\": 20}\n",
    "    ]\n",
    "    result = all_toxin_data_collection.aggregate(pipeline)\n",
    "    result = list(result)\n",
    "\n",
    "    df = pd.DataFrame(result)\n",
    "    df_str = df.to_string()\n",
    "\n",
    "    # Clear the text widget\n",
    "    text_widget.config(state=NORMAL)\n",
    "    text_widget.delete(\"1.0\", END)\n",
    "\n",
    "    # Insert the DataFrame string into the text widget\n",
    "    text_widget.insert(END, df_str)\n",
    "\n",
    "    # Make the text widget read-only\n",
    "    text_widget.config(state=DISABLED)\n",
    "\n",
    "def find_State():\n",
    "    global find_window,dropdown,text_widget\n",
    "    find_window = Toplevel(root)\n",
    "    find_window.title(\"Find Window\")\n",
    "    find_window.geometry(\"700x500\")\n",
    "    \n",
    "    find_window.configure(background=\"lavender\")\n",
    "    \n",
    "    state = ['Solid', 'Gas', 'Liquid']\n",
    "    dropdown = Combobox(find_window, values=state)\n",
    "    dropdown.pack()\n",
    "\n",
    "    # Set the default selected value\n",
    "    dropdown.set(state[0])\n",
    "\n",
    "    show_button = Button(find_window, text=\"Show\",width=20,height=2,bg=\"plum4\" ,command=show_dataframe)\n",
    "    show_button.pack()\n",
    "\n",
    "    text_widget = Text(find_window)\n",
    "    text_widget.pack()\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find query based on molecular weight"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_database():\n",
    "    v = float(entry.get())\n",
    "\n",
    "    col = db[\"AllToxinTarget\"]\n",
    "    documents = col.find({\"Molecular Weight\": {\"$gt\": v}}, {\"_id\": 0, \"NCBI Taxonomy Id\": 1, \"UniProt Name\": 1})\n",
    "    result = list(documents)\n",
    "    df = pd.DataFrame(result)\n",
    "\n",
    "    # Clear the text widget\n",
    "    text_widget.config(state=NORMAL)\n",
    "    text_widget.delete(\"1.0\", END)\n",
    "\n",
    "    # Insert the DataFrame string into the text widget\n",
    "    text_widget.insert(END, df.to_string())\n",
    "\n",
    "    # Make the text widget read-only\n",
    "    text_widget.config(state=DISABLED)\n",
    "\n",
    "def find_molecule():\n",
    "    global molecule_window, entry, text_widget\n",
    "\n",
    "    molecule_window = Toplevel(root)\n",
    "    molecule_window.title(\"Find Window\")\n",
    "    molecule_window.geometry(\"700x500\")\n",
    "    molecule_window.configure(background=\"lavender\")\n",
    "\n",
    "    entry_label = Label(molecule_window, text=\"Enter Molecular Weight:\")\n",
    "    entry_label.pack()\n",
    "\n",
    "    entry = Entry(molecule_window)\n",
    "    entry.pack()\n",
    "\n",
    "    show_button = Button(molecule_window, text=\"Show\", width=20, height=2, bg=\"plum4\", command=search_database)\n",
    "    show_button.pack()\n",
    "\n",
    "    text_widget = Text(molecule_window)\n",
    "    text_widget.pack()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find query based on common name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_database_commonName():\n",
    "    v = entry.get()\n",
    "\n",
    "    col = db['AllToxinData']\n",
    "    documents = col.find({\"common_name\":v},{\"_id\":0,'description':1})\n",
    "    result = list(documents)\n",
    "    df = pd.DataFrame(result)\n",
    "\n",
    "    # Clear the text widget\n",
    "    text_widget.config(state=NORMAL)\n",
    "    text_widget.delete(\"1.0\", END)\n",
    "\n",
    "    # Insert the DataFrame string into the text widget\n",
    "    text_widget.insert(END, df.to_string())\n",
    "\n",
    "    # Make the text widget read-only\n",
    "    text_widget.config(state=DISABLED)\n",
    "\n",
    "def find_commonName():\n",
    "    global common_name_Window, entry, text_widget\n",
    "\n",
    "    common_name_Window = Toplevel(root)\n",
    "    common_name_Window.title(\"Find bassed on common Name\")\n",
    "    common_name_Window.geometry(\"700x500\")\n",
    "    common_name_Window.configure(background=\"lavender\")\n",
    "\n",
    "    entry_label = Label(common_name_Window, text=\"Enter Molecular Weight:\")\n",
    "    entry_label.pack()\n",
    "\n",
    "    entry = Entry(common_name_Window)\n",
    "    entry.pack()\n",
    "\n",
    "    show_button = Button(common_name_Window, text=\"Show\", width=20, height=2, bg=\"plum4\", command=search_database_commonName)\n",
    "    show_button.pack()\n",
    "\n",
    "    text_widget = Text(common_name_Window)\n",
    "    text_widget.pack()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### query based on gene bank id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_target_viewer():\n",
    "    # Create toxin target sequence viewer window\n",
    "    viewer_window = Toplevel(root)\n",
    "    viewer_window.title(\"Toxin Target Sequence Viewer\")\n",
    "    viewer_window.geometry(\"800x700\")  # Increase the window size\n",
    "    viewer_window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Define the function to execute the query and display the result in a text box\n",
    "    def execute_query():\n",
    "        gene_id = input_text.get()\n",
    "        documents = all_toxin_target_collection.find({\"Genbank Gene ID\": gene_id}, {\"_id\": 0, 'Gene Sequence': 1, 'Amino Acid Sequence': 1})\n",
    "        result = list(documents)\n",
    "        df = pd.DataFrame(result)\n",
    "        result_text.configure(state='normal')\n",
    "        result_text.delete('1.0', tk.END)\n",
    "        result_text.insert(END, df.to_string(index=False))\n",
    "        result_text.configure(state='disabled')\n",
    "\n",
    "    # Create input label and text box\n",
    "    input_label = Label(viewer_window, text=\"Enter Genbank Gene ID:\", font=(\"Arial\", 12), bg=\"#C88EA7\")\n",
    "    input_label.pack(pady=10)\n",
    "    input_text = Entry(viewer_window, font=(\"Arial\", 12), width=40)\n",
    "    input_text.pack(pady=10)\n",
    "\n",
    "    # Create button to execute query\n",
    "    execute_button = Button(viewer_window, text=\"Execute Query\", command=execute_query)\n",
    "    execute_button.pack(pady=10)\n",
    "\n",
    "    # Create text box to display result\n",
    "    result_text = Text(viewer_window, font=(\"Arial\", 12), bg=\"white\", width=100, height=25, wrap=tk.WORD)  # Increase the height\n",
    "    result_text.pack(pady=10)\n",
    "\n",
    "    # Create close button\n",
    "    close_button = Button(viewer_window, text=\"Close\",width=25,height=3 ,command=viewer_window.destroy)\n",
    "    close_button.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "\n",
    "def open_data_query_window():\n",
    "\n",
    "    # Create the main window\n",
    "    find = Toplevel(root)\n",
    "    find.title(\"Data Querying\")\n",
    "    find.configure(bg='steelblue4')\n",
    "\n",
    "    def open_data_query_window():\n",
    "        # Function to open the data query window\n",
    "        window_width = 300\n",
    "        window_height = 200\n",
    "        center_window(find, window_width, window_height)\n",
    "\n",
    "        button_frame = tk.Frame(find, bg='steelblue4')\n",
    "        button_frame.pack(pady=50)\n",
    "\n",
    "        find_toxin_data_button = tk.Button(button_frame, text=\"Find All Toxin Data\",\n",
    "                                          width=25, height=2, command=open_find_queries_window)\n",
    "        find_toxin_data_button.pack(pady=10)\n",
    "\n",
    "        find_target_data_button = tk.Button(button_frame, text=\"Find All Target Data\",\n",
    "                                           width=25, height=2, command=open_find_queries_window)\n",
    "        find_target_data_button.pack(pady=10)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_data_find_window():\n",
    "    # Create the Toplevel window\n",
    "    top_window = tk.Toplevel(root)\n",
    "    top_window.title(\"Toplevel Window\")\n",
    "    top_window.configure(bg='steelblue4')\n",
    "    \n",
    "    window_width = 400\n",
    "    window_height = 300\n",
    "    top_window.geometry(f\"{window_width}x{window_height}\")\n",
    "\n",
    "    \n",
    "    # Create Button 1\n",
    "    button1 = tk.Button(top_window, text=\"Find All Toxin Data\", width=25, height=2, command=open_find_queries_window)\n",
    "    button1.pack(pady=10)\n",
    "\n",
    "    # Create Button 2\n",
    "    button2 = tk.Button(top_window, text=\"Find All Target Data\", width=25, height=2, command=open_find_queries_window_target)\n",
    "    button2.pack(pady=10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### query base on common name indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_insertion_function():\n",
    "    # Create the Toplevel window\n",
    "    top_window = tk.Toplevel(root)\n",
    "    top_window.title(\"Data Insertion\")\n",
    "    top_window.configure(bg='steelblue4')\n",
    "    \n",
    "    # Set the width and height of the window\n",
    "    window_width = 400\n",
    "    window_height = 300\n",
    "    top_window.geometry(f\"{window_width}x{window_height}\")\n",
    "\n",
    "    # Create Button 1\n",
    "    button1 = tk.Button(top_window, text=\"toxins data\", width=25, height=2, command=dataentry)\n",
    "    button1.pack(pady=10)\n",
    "\n",
    "    # Create Button 2\n",
    "    button2 = tk.Button(top_window, text=\"target data\", width=25, height=2, command=insert_toxinTarget)\n",
    "    button2.pack(pady=10)\n",
    "    \n",
    "    # Create Button 3\n",
    "    button3 = tk.Button(top_window, text=\"close\", width=25, height=2, command=top_window.destroy)\n",
    "    button3.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### delete query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "from tkinter import messagebox\n",
    "from pymongo import MongoClient\n",
    "\n",
    "# Connect to MongoDB\n",
    "client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "db = client[\"toxins\"]\n",
    "all_toxin_target_collection = db[\"AllToxinTarget\"]\n",
    "\n",
    "def delete_record_by_molecular_weight():\n",
    "    # Create a new window for the GUI\n",
    "    window = tk.Toplevel(root)\n",
    "    window.title(\"Delete Record by ID\")\n",
    "    window.geometry(\"300x150\")\n",
    "    window.configure(bg='white')\n",
    "\n",
    "    # Function to handle the deletion\n",
    "    def delete_record():\n",
    "        molecular_weight = entry.get()\n",
    "\n",
    "        # Construct the query\n",
    "        query = {\"Molecular Weight\": {\"$lt\": float(molecular_weight)}}\n",
    "\n",
    "        # Perform the deletion\n",
    "        result = all_toxin_target_collection.delete_many(query)\n",
    "\n",
    "        # Display the result in a message box\n",
    "        if result.deleted_count > 0:\n",
    "            messagebox.showinfo(\"Deletion Result\", f\"{result.deleted_count} record(s) deleted successfully!\")\n",
    "        else:\n",
    "            messagebox.showinfo(\"Deletion Result\", \"No matching record found.\")\n",
    "\n",
    "    # Create the GUI elements\n",
    "    label = tk.Label(window, text=\"Enter the Molecular Weight:\")\n",
    "    label.pack(pady=10)\n",
    "\n",
    "    entry = tk.Entry(window)\n",
    "    entry.pack(pady=5)\n",
    "\n",
    "    delete_button = tk.Button(window, text=\"Delete\", command=delete_record)\n",
    "    delete_button.pack(pady=10)\n",
    "\n",
    "def delete_one_by_molecular_weight():\n",
    "    # Create a new window for the GUI\n",
    "    window = tk.Toplevel(root)\n",
    "    window.title(\"Delete Record by Molecular Weight\")\n",
    "    window.geometry(\"300x150\")\n",
    "    window.configure(bg='white')\n",
    "\n",
    "    # Function to handle the deletion\n",
    "    def delete_record():\n",
    "        molecular_weight = entry.get()\n",
    "\n",
    "        # Construct the query\n",
    "        query = {\"_id\": molecular_weight}\n",
    "\n",
    "        # Perform the deletion\n",
    "        result = all_toxin_data_collection.delete_one(query)\n",
    "\n",
    "        # Display the result in a message box\n",
    "        if result.deleted_count > 0:\n",
    "            messagebox.showinfo(\"Deletion Result\", f\"{result.deleted_count} record(s) deleted successfully!\")\n",
    "        else:\n",
    "            messagebox.showinfo(\"Deletion Result\", \"No matching record found.\")\n",
    "\n",
    "    # Create the GUI elements\n",
    "    label = tk.Label(window, text=\"Enter the Molecular Weight:\")\n",
    "    label.pack(pady=10)\n",
    "\n",
    "    entry = tk.Entry(window)\n",
    "    entry.pack(pady=5)\n",
    "\n",
    "    delete_button = tk.Button(window, text=\"Delete\", command=delete_record)\n",
    "    delete_button.pack(pady=10) \n",
    "import tkinter as tk\n",
    "from tkinter import messagebox\n",
    "from pymongo import MongoClient\n",
    "\n",
    "# Connect to MongoDB\n",
    "client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "db = client[\"toxins\"]\n",
    "all_toxin_target_collection = db[\"AllToxinTarget\"]\n",
    "\n",
    "def delete_one_record():\n",
    "    # Function to handle deleting one record\n",
    "    # Add your code here for deleting one record\n",
    "    pass\n",
    "\n",
    "def delete_many_records():\n",
    "    # Function to handle deleting many records\n",
    "    # Add your code here for deleting many records\n",
    "    pass\n",
    "\n",
    "def delete_window():\n",
    "    # Create the delete window\n",
    "    window = tk.Toplevel()\n",
    "    window.title(\"Delete Records\")\n",
    "    window.geometry(\"300x150\")\n",
    "    window.configure(bg='steelblue4')\n",
    "\n",
    "    # Create the buttons\n",
    "    delete_one_button = tk.Button(window, text=\"Delete One\", width=15, command=delete_one_by_molecular_weight)\n",
    "    delete_one_button.pack(pady=10)\n",
    "\n",
    "    delete_many_button = tk.Button(window, text=\"Delete Many\", width=15, command=delete_record_by_molecular_weight)\n",
    "    delete_many_button.pack(pady=10)\n",
    "\n",
    "    close_button = tk.Button(window, text=\"Close\", width=15, command=window.destroy)\n",
    "    close_button.pack(pady=10)\n",
    "\n",
    "    window.mainloop()\n",
    "\n",
    "def open_delete_window():\n",
    "    # Function to open the delete window\n",
    "    delete_window()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tkinter as tk\n",
    "\n",
    "def query_window():\n",
    "    \n",
    "    # Create the main window\n",
    "    open_window =Toplevel(root)\n",
    "    open_window.title(\"data_quering\")\n",
    "    open_window.geometry(\"400x300\")\n",
    "    open_window.configure(bg='steelblue4')\n",
    "\n",
    "    # Create a frame to hold the buttons\n",
    "    button_frame = tk.Frame(open_window, bg='steelblue4')\n",
    "    button_frame.pack(pady=50)\n",
    "\n",
    "    # Create the buttons\n",
    "    insert_button = tk.Button(button_frame, text=\"Insert\", width=10, bg='grey', command=data_insertion_function)\n",
    "    insert_button.pack(side=tk.LEFT, padx=10)\n",
    "\n",
    "    update_button = tk.Button(button_frame, text=\"Update\", width=10, bg='grey', command=open_update_window_entry)\n",
    "    update_button.pack(side=tk.LEFT, padx=10)\n",
    "\n",
    "    delete_button = tk.Button(button_frame, text=\"Delete\", width=10, bg='grey', command=open_delete_window)\n",
    "    delete_button.pack(side=tk.LEFT, padx=10)\n",
    "\n",
    "    find_button = tk.Button(button_frame, text=\"Find\", width=10, bg='grey', command=open_data_find_window)\n",
    "    find_button.pack(side=tk.LEFT, padx=10)\n",
    "\n",
    "    close_button = tk.Button(open_window, text=\"Close\", width=10, bg='grey', command=open_window.destroy)\n",
    "    close_button.pack(pady=10)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "################################################################3\n",
    "#################################################################3\n",
    "3#################################################################3\n",
    "###############################################################3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_common_name_indexing():\n",
    "    # Create a new window for the Toxin Data Viewer\n",
    "    viewer_window = tk.Toplevel(root)\n",
    "    viewer_window.title(\"Toxin Data Viewer\")\n",
    "    viewer_window.configure(bg=\"#C88EA7\")\n",
    "    \n",
    "    # Define function to execute the query and display the result in a data frame\n",
    "    def execute_query():\n",
    "        sorting_order = 1 if sorting_radio.get() == 1 else -1\n",
    "        result = all_toxin_data_collection.find({}, {\"chemical_formula\": 1, \"state\": 1}).sort(\"common_name\", sorting_order).limit(10)\n",
    "        df = pd.DataFrame(list(result))\n",
    "\n",
    "        # Display the result in a data frame\n",
    "        result_text.configure(state='normal')\n",
    "        result_text.delete('1.0', tk.END)\n",
    "        result_text.insert(tk.END, df.to_string(index=False))\n",
    "        result_text.configure(state='disabled')\n",
    "    \n",
    "    # Create sorting order radio buttons\n",
    "    sorting_radio = tk.IntVar()\n",
    "    ascending_radio = tk.Radiobutton(viewer_window, text=\"Ascending\", variable=sorting_radio, value=1)\n",
    "    ascending_radio.pack()\n",
    "    descending_radio = tk.Radiobutton(viewer_window, text=\"Descending\", variable=sorting_radio, value=-1)\n",
    "    descending_radio.pack()\n",
    "\n",
    "    # Create button to execute query\n",
    "    execute_button = tk.Button(viewer_window, text=\"Execute Query\", command=execute_query)\n",
    "    execute_button.pack(pady=10)\n",
    "\n",
    "    # Create text box to display result\n",
    "    result_text = tk.Text(viewer_window, font=(\"Arial\", 12), bg=\"white\", width=60, height=10, wrap=tk.WORD)\n",
    "    result_text.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_find_queries_window_target():\n",
    "    find_queries_window_target = Toplevel(root)\n",
    "    find_queries_window_target.title(\"Find Queries - target data\")\n",
    "    find_queries_window_target.geometry(\"600x400\")\n",
    "\n",
    "    # Set background image\n",
    "    background_image = PhotoImage(file=\"find.PNG\")\n",
    "    background_label = Label(find_queries_window_target, image=background_image)\n",
    "    background_label.place(x=0, y=0, relwidth=1, relheight=1)\n",
    "\n",
    "    # Create buttons\n",
    "    find_mol = Button(find_queries_window_target, text=\"Find Molecule\", width=20, bg=\"#9c8d91\", fg=\"white\", command=find_molecule)\n",
    "    find_based_on_index = Button(find_queries_window_target, text=\"Find Based on Index\", width=15, height=2, bg=\"#9c8d91\", fg=\"white\", command=indexing_find)\n",
    "    button3 = Button(find_queries_window_target, text=\"find commonName Data\", width=20, bg=\"#9c8d91\", fg=\"white\", command=open_toxin_viewer)\n",
    "    find_gene_id = Button(find_queries_window_target, text=\"genebankID\", width=20, bg=\"#9c8d91\", fg=\"white\", command=open_target_viewer)\n",
    "    button6 = Button(find_queries_window_target, text=\"common_name indexing\", width=20, bg=\"#9c8d91\", fg=\"white\", command=open_common_name_indexing)\n",
    "    button5 = Button(find_queries_window_target, text=\"close\", width=20, bg=\"#9c8d91\", fg=\"white\", command=find_queries_window_target.destroy)\n",
    "\n",
    "    # Grid placement of buttons\n",
    "    find_mol.grid(row=0, column=0, padx=10, pady=10)\n",
    "    find_based_on_index.grid(row=1, column=0, padx=10, pady=10)\n",
    "    button3.grid(row=2, column=0, padx=10, pady=10)\n",
    "    find_gene_id.grid(row=3, column=0, padx=10, pady=10)\n",
    "    button6.grid(row=4, column=0, padx=10, pady=10)\n",
    "    button5.grid(row=5, column=0, padx=10, pady=10)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find common name data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_toxin_viewer():\n",
    "    # Create toxin information viewer window\n",
    "    viewer_window = tk.Toplevel(root)\n",
    "    viewer_window.title(\"Toxin Information Viewer\")\n",
    "    viewer_window.geometry(\"800x500\")  # Increase the window size\n",
    "    viewer_window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Define the function to execute the query and display the result in a text box\n",
    "    def execute_query():\n",
    "        common_name = input_text.get()\n",
    "        documents = all_toxin_data_collection.find({\"common_name\": common_name}, {\"_id\": 0, \"toxicity\": 1, \"carcinogenicity\": 1, \"treatment\": 1})\n",
    "        result = list(documents)\n",
    "        df = pd.DataFrame(result)\n",
    "        result_text.configure(state='normal')\n",
    "        result_text.delete('1.0', tk.END)\n",
    "        result_text.insert(tk.END, df.to_string(index=False))\n",
    "        result_text.configure(state='disabled')\n",
    "\n",
    "    # Create input label and text box\n",
    "    input_label = tk.Label(viewer_window, text=\"Enter Common Name:\", font=(\"Arial\", 12), bg=\"#C88EA7\")\n",
    "    input_label.pack(pady=10)\n",
    "    input_text = tk.Entry(viewer_window, font=(\"Arial\", 12), width=40)\n",
    "    input_text.pack(pady=10)\n",
    "\n",
    "    # Create button to execute query\n",
    "    execute_button = tk.Button(viewer_window, text=\"Execute Query\", command=execute_query)\n",
    "    execute_button.pack(pady=10)\n",
    "\n",
    "    # Create text box to display result\n",
    "    result_text = tk.Text(viewer_window, font=(\"Arial\", 12), bg=\"white\", width=100, height=20, wrap=tk.WORD)  # Increase the width\n",
    "    result_text.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### search based on toxin id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_toxin_searcher():\n",
    "    # Create toxin information searcher window\n",
    "    searcher_window = tk.Toplevel(root)\n",
    "    searcher_window.title(\"Toxin Information Searcher\")\n",
    "    searcher_window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Define the function to search for data in the database\n",
    "    def search_data():\n",
    "        search_id = id_entry.get()\n",
    "        documents = all_toxin_data_collection.find({'_id': search_id}, {'_id': 0, 'metabolism': 1, \"cellular_locations\": 1,\"Locus\":1})\n",
    "        result = list(documents)\n",
    "        df = pd.DataFrame(result)\n",
    "        result_text.configure(state='normal')\n",
    "        result_text.delete('1.0', tk.END)\n",
    "        result_text.insert(tk.END, df.to_string(index=False))\n",
    "        result_text.configure(state='disabled')\n",
    "\n",
    "    # Create input label and text box\n",
    "    id_label = tk.Label(searcher_window, text=\"Enter ID to search for:\", font=(\"Arial\", 12), bg=\"#C88EA7\")\n",
    "    id_label.pack(pady=10)\n",
    "    id_entry = tk.Entry(searcher_window, font=(\"Arial\", 12), width=40)\n",
    "    id_entry.pack(pady=10)\n",
    "\n",
    "    # Create search button to trigger the search function\n",
    "    search_button = tk.Button(searcher_window, text=\"Search\", font=(\"Arial\", 12), command=search_data)\n",
    "    search_button.pack(pady=10)\n",
    "\n",
    "    # Create result label and text box\n",
    "    result_label = tk.Label(searcher_window, text=\"Search Results:\", font=(\"Arial\", 12), bg=\"#C88EA7\")\n",
    "    result_label.pack(pady=10)\n",
    "    result_text = tk.Text(searcher_window, font=(\"Arial\", 12), width=80, height=20, state='disabled')\n",
    "    result_text.pack(pady=10)\n",
    "\n",
    "    # Create close button\n",
    "    close_button = tk.Button(searcher_window, text=\"Close\", command=searcher_window.destroy)\n",
    "    close_button.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find minumum id moldb weight"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def execute_pipeline():\n",
    "    pipeline = [\n",
    "        {\n",
    "            \"$match\": {\n",
    "                \"moldb_mono_mass\": {\n",
    "                    \"$ne\": float(\"nan\"),\n",
    "                    \"$type\": \"double\"\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"$group\": {\n",
    "                \"_id\": None,\n",
    "                \"min_value\": {\n",
    "                    \"$min\": \"$moldb_mono_mass\"\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "    result = all_toxin_data_collection.aggregate(pipeline)\n",
    "    min_value = result.next()[\"min_value\"]\n",
    "    d = all_toxin_data_collection.find_one({'moldb_mono_mass': min_value})['_id']\n",
    "    data = [\n",
    "        {'_id': d, 'minimum value': min_value}\n",
    "    ]\n",
    "    df = pd.DataFrame(data)\n",
    "    messagebox.showinfo(\"Result\", df.to_string(index=False))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find minimum weight  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_value = None\n",
    "max_value = None\n",
    "\n",
    "def find_min_weight():\n",
    "    pipeline = [\n",
    "        {\n",
    "            \"$match\": {\n",
    "                \"weight\": { \"$regex\": \"^[0-9\\.]+$\" }\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"$addFields\": {\n",
    "                \"floatField\": {\n",
    "                    \"$convert\": {\n",
    "                        \"input\": \"$weight\",\n",
    "                        \"to\": \"double\"\n",
    "                    }\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"$group\": {\n",
    "                \"_id\": None,\n",
    "                \"minValue\": { \"$min\": \"$floatField\" }\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "\n",
    "    result = all_toxin_data_collection.aggregate(pipeline)\n",
    "    global min_value\n",
    "    min_value = result.next()[\"minValue\"]\n",
    "\n",
    "def find_max_weight():\n",
    "    pipeline = [\n",
    "        {\n",
    "            \"$match\": {\n",
    "                \"weight\": { \"$regex\": \"^[0-9\\.]+$\" }\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"$addFields\": {\n",
    "                \"floatField\": {\n",
    "                    \"$convert\": {\n",
    "                        \"input\": \"$weight\",\n",
    "                        \"to\": \"double\"\n",
    "                    }\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"$group\": {\n",
    "                \"_id\": None,\n",
    "                \"maxValue\": { \"$max\": \"$floatField\" }\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "\n",
    "    result = all_toxin_data_collection.aggregate(pipeline)\n",
    "    global max_value\n",
    "    max_value = result.next()[\"maxValue\"]\n",
    "\n",
    "def show_message_box(value_type, value):\n",
    "    messagebox.showinfo(f\"Maximum {value_type}\", f\"The {value_type} value is: {value}\")\n",
    "\n",
    "def open_weight_window():\n",
    "    # Create a new window for weight operations\n",
    "    weight_window = tk.Toplevel(root)\n",
    "    weight_window.title(\"Weight Operations\")\n",
    "    weight_window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Create button to find minimum weight\n",
    "    min_button = tk.Button(weight_window, text=\"calculate Minimum Weight\", command=find_min_weight)\n",
    "    min_button.pack(pady=10)\n",
    "\n",
    "    # Create button to find maximum weight\n",
    "    max_button = tk.Button(weight_window, text=\"calculate Maximum Weight\", command=find_max_weight)\n",
    "    max_button.pack(pady=10)\n",
    "\n",
    "    # Create button to show minimum weight\n",
    "    min_value_button = tk.Button(weight_window, text=\"Show Minimum Weight\", command=lambda: show_message_box(\"Minimum Weight\", min_value))\n",
    "    min_value_button.pack(pady=10)\n",
    "\n",
    "    # Create button to show maximum weight\n",
    "    max_value_button = tk.Button(weight_window, text=\"Show Maximum Weight\", command=lambda: show_message_box(\"Maximum Weight\", max_value))\n",
    "    max_value_button.pack(pady=10)\n",
    "\n",
    "    # Create close button for the weight operations window\n",
    "    close_button = tk.Button(weight_window, text=\"Close\", command=weight_window.destroy)\n",
    "    close_button.pack(pady=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define function to handle button click event\n",
    "def handle_button_click():\n",
    "    x = all_toxin_data_collection.find({} , {\"moldb_mono_mass\":1}).sort(\"moldb_mono_mass\",-1).limit(1)\n",
    "    x = list(x)\n",
    "    df = pd.DataFrame(x)\n",
    "    result = str(df.at[0, \"moldb_mono_mass\"])\n",
    "    messagebox.showinfo(\"Maximum Moldb Mono Mass\", result)\n",
    "\n",
    "def open_window():\n",
    "    # Create a new window for the toxin data collection\n",
    "    window = tk.Toplevel(root)\n",
    "    window.title(\"Toxin Data Collection\")\n",
    "    window.geometry(\"400x300\")\n",
    "    window.configure(bg=\"light green\")\n",
    "\n",
    "    # Create button to find the maximum Moldb Mono Mass\n",
    "    find_button = tk.Button(window, text=\"Find Maximum Moldb Mono Mass\", command=handle_button_click)\n",
    "    find_button.pack()\n",
    "    \n",
    "    find_id = tk.Button(window, text=\"Find minimumMoldb_id\", command=open_execute_window)\n",
    "    find_id.pack()\n",
    "    \n",
    "    \n",
    "    close_btn=Button(window, text=\"Close\", command=window.destroy)\n",
    "    close_btn.pack()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find toxin type "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "import pymongo\n",
    "\n",
    "def open_toxin_data_collection_gui():\n",
    "    # Connect to MongoDB\n",
    "\n",
    "\n",
    "    # Define function to handle button click event\n",
    "    def handle_button_click():\n",
    "        toxin_id = toxin_id_entry.get()\n",
    "        documents = all_toxin_data_collection.find({\"_id\": toxin_id})\n",
    "        toxin_types = \"\"\n",
    "        for document in documents:\n",
    "            toxin_types += document[\"toxin_structure\"][\"TYPES\"] + \"\\n\"\n",
    "        toxin_types_text.delete(1.0, tk.END) # Clear previous result\n",
    "        toxin_types_text.insert(tk.END, toxin_types) # Update text widget with new result\n",
    "\n",
    "    # Create GUI window\n",
    "    window = tk.Tk()\n",
    "    window.title(\"Toxin Data Collection\")\n",
    "    window.geometry(\"400x300\")\n",
    "    window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Create GUI widgets\n",
    "    toxin_id_label = tk.Label(window, text=\"Enter Toxin ID:\")\n",
    "    toxin_id_entry = tk.Entry(window)\n",
    "    submit_button = tk.Button(window, text=\"Submit\", command=handle_button_click)\n",
    "    toxin_types_label = tk.Label(window, text=\"Toxin Types:\")\n",
    "    toxin_types_text = tk.Text(window, height=10, width=30)\n",
    "\n",
    "    # Add widgets to GUI window\n",
    "    toxin_id_label.pack()\n",
    "    toxin_id_entry.pack()\n",
    "    submit_button.pack()\n",
    "    toxin_types_label.pack()\n",
    "    toxin_types_text.pack()\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find toxins smiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "import pymongo\n",
    "\n",
    "def open_toxin_data_smile_gui():\n",
    "    # Connect to MongoDB\n",
    "    client = pymongo.MongoClient(\"mongodb://localhost:27017/\")\n",
    "    db = client[\"toxins\"]\n",
    "    all_toxin_data_collection = db[\"AllToxinData\"]\n",
    "\n",
    "    # Define function to handle button click event\n",
    "    def handle_button_click():\n",
    "        toxin_id = toxin_id_entry.get()\n",
    "        documents = all_toxin_data_collection.find({\"_id\": toxin_id})\n",
    "        toxin_smiles = \"\"\n",
    "        for document in documents:\n",
    "            toxin_smiles += document[\"toxin_structure\"][\"SMILES\"] + \"\\n\"\n",
    "        toxin_smiles_text.delete(1.0, tk.END) # Clear previous result\n",
    "        toxin_smiles_text.insert(tk.END, toxin_smiles) # Update text widget with new result\n",
    "\n",
    "    # Create GUI window\n",
    "    window = Toplevel(root)\n",
    "    window.title(\"Toxin Data Collection\")\n",
    "    window.geometry(\"400x300\")\n",
    "    window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Create GUI widgets\n",
    "    toxin_id_label = tk.Label(window, text=\"Enter Toxin ID:\")\n",
    "    toxin_id_entry = tk.Entry(window)\n",
    "    submit_button = tk.Button(window, text=\"Submit\", command=handle_button_click)\n",
    "    toxin_smiles_label = tk.Label(window, text=\"Toxin SMILES:\")\n",
    "    toxin_smiles_text = tk.Text(window, height=10, width=30)\n",
    "\n",
    "    # Add widgets to GUI window\n",
    "    toxin_id_label.pack()\n",
    "    toxin_id_entry.pack()\n",
    "    submit_button.pack()\n",
    "    toxin_smiles_label.pack()\n",
    "    toxin_smiles_text.pack()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find toxins chem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "import pymongo\n",
    "\n",
    "def open_toxin_data_chem_gui():\n",
    "\n",
    "    # Define function to handle button click event\n",
    "    def handle_button_click():\n",
    "        toxin_id = toxin_id_entry.get()\n",
    "        documents = all_toxin_data_collection.find({\"_id\": toxin_id})\n",
    "        toxin_smiles = \"\"\n",
    "        for document in documents:\n",
    "            toxin_smiles += document[\"toxin_structure\"][\"SMILES\"] + \"\\n\"\n",
    "        toxin_smiles_text.delete(1.0, tk.END) # Clear previous result\n",
    "        toxin_smiles_text.insert(tk.END, toxin_smiles) # Update text widget with new result\n",
    "\n",
    "    # Create GUI window\n",
    "    window = Toplevel(root)\n",
    "    window.title(\"Toxin Data Collection\")\n",
    "    window.geometry(\"400x300\")\n",
    "    window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Create GUI widgets\n",
    "    toxin_id_label = tk.Label(window, text=\"Enter Toxin ID:\")\n",
    "    toxin_id_entry = tk.Entry(window)\n",
    "    submit_button = tk.Button(window, text=\"Submit\", command=handle_button_click)\n",
    "    toxin_smiles_label = tk.Label(window, text=\"Toxin SMILES:\")\n",
    "    toxin_smiles_text = tk.Text(window, height=10, width=30)\n",
    "\n",
    "    # Add widgets to GUI window\n",
    "    toxin_id_label.pack()\n",
    "    toxin_id_entry.pack()\n",
    "    submit_button.pack()\n",
    "    toxin_smiles_label.pack()\n",
    "    toxin_smiles_text.pack()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### find data  by molecular weight"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "import pandas as pd\n",
    "from pymongo import MongoClient\n",
    "\n",
    "# Connect to MongoDB\n",
    "\n",
    "# Define function to open the window\n",
    "def open_toxin_molecular_weight_viewer():\n",
    "    # Create tkinter GUI window\n",
    "    window = Toplevel(root)\n",
    "    window.title(\"Toxin Molecular Weight Viewer\")\n",
    "    window.configure(bg=\"#C88EA7\")\n",
    "\n",
    "    # Define the function to execute the query and display the result in a text box\n",
    "    def execute_query():\n",
    "        molecular_weight = float(input_text.get())\n",
    "        documents = all_toxin_target_collection.find({\"Molecular Weight\": {\"$gt\": molecular_weight}}, {\"_id\": 0, 'NCBI Taxonomy Id': 1, 'UniProt Name': 1})\n",
    "        result = list(documents)\n",
    "        df = pd.DataFrame(result)\n",
    "        result_text.configure(state='normal')\n",
    "        result_text.delete('1.0', tk.END)\n",
    "        result_text.insert(tk.END, df.to_string(index=False))\n",
    "        result_text.configure(state='disabled')\n",
    "\n",
    "    # Create input label and text box\n",
    "    input_label = tk.Label(window, text=\"Enter id:\", font=(\"Arial\", 12), bg=\"#C88EA7\")\n",
    "    input_label.pack(pady=10)\n",
    "    input_text = tk.Entry(window, font=(\"Arial\", 12), width=40)\n",
    "    input_text.pack(pady=10)\n",
    "\n",
    "    # Create button to execute query\n",
    "    execute_button = tk.Button(window, text=\"Execute Query\", command=execute_query)\n",
    "    execute_button.pack(pady=10)\n",
    "\n",
    "    # Create text box to display result\n",
    "    result_text = tk.Text(window, font=(\"Arial\", 12), bg=\"white\", width=60, height=10, wrap=tk.WORD)\n",
    "    result_text.pack(pady=10)\n",
    "\n",
    "\n",
    "# Define function to open the Toxin Molecular Weight Viewer window\n",
    "def open_molecular_weight_viewer_window():\n",
    "    open_toxin_molecular_weight_viewer()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_find_queries_window():\n",
    "    find_queries_window = tk.Toplevel(root)\n",
    "    find_queries_window.title(\"Find Queries - allToxin_data\")\n",
    "    find_queries_window.geometry(\"600x400\")\n",
    "\n",
    "    # Set background image\n",
    "    background_image = tk.PhotoImage(file=\"find.PNG\")\n",
    "    background_label = tk.Label(find_queries_window, image=background_image)\n",
    "    background_label.place(x=0, y=0, relwidth=1, relheight=1)\n",
    "\n",
    "    # Create buttons with the same height\n",
    "    button_height = 2\n",
    "\n",
    "    # Define the buttons in the desired order\n",
    "    buttons = [\n",
    "        (\"Find State\", find_State),\n",
    "        (\"Common Name Data\", open_toxin_viewer),\n",
    "        (\"Find Toxin ID\", open_toxin_searcher),\n",
    "        (\"Minimum Weight\", open_weight_window),\n",
    "        (\"Maximum Moldb Mono Mass\", open_window),\n",
    "        (\"Toxins SMILES\", open_toxin_data_smile_gui),\n",
    "        (\"Toxins Chem\", open_toxin_data_chem_gui),\n",
    "        (\"Find Toxin Type\", open_toxin_data_collection_gui)\n",
    "        ,(\"find_data_by_weight\",open_molecular_weight_viewer_window)\n",
    "    ]\n",
    "\n",
    "    # Calculate the number of rows and columns for button placement\n",
    "    num_rows = 3\n",
    "    num_columns = (len(buttons) + num_rows - 1) // num_rows\n",
    "\n",
    "    # Create and place buttons in the window\n",
    "    for index, (button_text, button_command) in enumerate(buttons):\n",
    "        row = index // num_columns\n",
    "        column = index % num_columns\n",
    "\n",
    "        button = tk.Button(find_queries_window, text=button_text, width=20, height=button_height, bg=\"#9c8d91\", fg=\"white\", command=button_command)\n",
    "        button.place(relx=(column + 0.5) / num_columns, rely=(row + 0.3) / num_rows, anchor=\"center\")\n",
    "\n",
    "    # Create the close button\n",
    "    close_button = tk.Button(find_queries_window, text=\"Close\", width=15, height=button_height, bg=\"#9c8d91\", fg=\"white\", command=find_queries_window.destroy)\n",
    "    close_button.place(relx=0.5, rely=0.9, anchor=\"center\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def center_window(window):\n",
    "    window.update_idletasks()\n",
    "    width = window.winfo_width()\n",
    "    height = window.winfo_height()\n",
    "    x = (window.winfo_screenwidth() // 2) - (width // 2)\n",
    "    y = (window.winfo_screenheight() // 2) - (height // 2)\n",
    "    window.geometry('{}x{}+{}+{}'.format(width, height, x, y))\n",
    "\n",
    "def Queries():\n",
    "    global Querywindow\n",
    "    Querywindow = Toplevel(root)\n",
    "    Querywindow.title(\"Query Window\")\n",
    "    Querywindow.geometry(\"800x500\")\n",
    "    center_window(Querywindow)\n",
    "    background_image = PhotoImage(file=\"query.PNG\")\n",
    "    background_label = Label(Querywindow, image=background_image)\n",
    "    background_label.image = background_image  # Retain a reference to the image\n",
    "    background_label.pack()\n",
    "    \n",
    "    ##############Buttons######################\n",
    "    \n",
    "    button_width = 200\n",
    "    button_height = 50\n",
    "\n",
    "\n",
    "    \n",
    "    find_mol = Button(Querywindow, text=\"Find Molecule\", bg=\"gold\", command=find_molecule)\n",
    "    find_mol.pack()\n",
    "    find_mol.place(relx=0.5, rely=0.55, anchor=CENTER, width=button_width, height=button_height)\n",
    "    \n",
    "    \n",
    "    find_based_on_index = Button(Querywindow, text=\"Find Based on Index\", bg=\"gold\",command=indexing_find)\n",
    "    find_based_on_index.pack()\n",
    "    find_based_on_index.place(relx=0.5, rely=0.75, anchor=CENTER, width=button_width, height=button_height)\n",
    "\n",
    "    close_button = Button(Querywindow, text=\"Close\", width=10, command=close_query)\n",
    "    close_button.pack()\n",
    "    close_button.place(relx=0.5, rely=0.85, anchor=CENTER, width=button_width, height=button_height)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# all toxins target insertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_toxinTarget():\n",
    "    global updateWindow\n",
    "\n",
    "    def update_data():\n",
    "        # Retrieve the input values from the entry fields\n",
    "        old_value = entry_old_value.get()\n",
    "        new_value = entry_new_value.get()\n",
    "\n",
    "        # Perform the update operation in the MongoDB collection\n",
    "        try:\n",
    "            # Update with your MongoDB connection details\n",
    "            db = client['toxins']  # Replace 'your_database' with your database name\n",
    "            collection = db[\"AllToxinTarget\"]  # Replace 'your_collection' with your collection name\n",
    "            collection.update_one({\"Name\": old_value}, {\"$set\": {\"Name\": new_value}})\n",
    "            messagebox.showinfo(\"Success\", \"Data updated successfully!\")\n",
    "        except Exception as e:\n",
    "            messagebox.showerror(\"Error\",str(e))\n",
    "\n",
    "    updateWindow = Toplevel(root)\n",
    "    updateWindow.title(\"Update Data\")\n",
    "    updateWindow.geometry(\"400x300\")\n",
    "    updateWindow.configure(bg=\"lavender\")  # Set background color to lavender\n",
    "\n",
    "    # Create labels and entry fields for old and new values\n",
    "    label_old_value = Label(updateWindow, text=\"Old Value:\")\n",
    "    label_old_value.pack()\n",
    "    entry_old_value = Entry(updateWindow)\n",
    "    entry_old_value.pack()\n",
    "\n",
    "    label_new_value = Label(updateWindow, text=\"New Value:\")\n",
    "    label_new_value.pack()\n",
    "    entry_new_value = Entry(updateWindow)\n",
    "    entry_new_value.pack()\n",
    "\n",
    "    # Create the update button\n",
    "    button_update = Button(updateWindow, text=\"Update\", command=update_data)\n",
    "    button_update.pack()\n",
    "\n",
    "    # Create the close button\n",
    "    button_close = Button(updateWindow, text=\"Close\", command=updateWindow.destroy)\n",
    "    button_close.pack()\n",
    "\n",
    "\n",
    "\n",
    "def open_targetWindow():\n",
    "    insert_toxinTarget()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_toxin_data_collection(var1: str, var2: str) -> pd.DataFrame:\n",
    "    # Connect to the MongoDB server and select the database and collection\n",
    "    client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "    db = client[\"toxins\"]\n",
    "    all_toxin_data_collection = db[\"AllToxinData\"]\n",
    "\n",
    "    # Update the documents with the old 'health_effects' value to the new value\n",
    "    myquery = {\"health_effects\": var1 }\n",
    "    newvalues = {\"$set\": {\"health_effects\": var2 } }\n",
    "    all_toxin_data_collection.update_one(myquery, newvalues)\n",
    "\n",
    "    # Find and return the documents with the new 'health_effects' value\n",
    "    documents = all_toxin_data_collection.find({\"health_effects\": var2},{\"health_effects\":1}).limit(3)\n",
    "    result = list(documents)\n",
    "    df = pd.DataFrame(result)\n",
    "    return df\n",
    "\n",
    "def open_update_window_health_effect():\n",
    "    # Create the top-level window for updating toxin data\n",
    "    update_window = Toplevel(root)\n",
    "    update_window.title(\"Update Toxin Data\")\n",
    "    update_window.configure(bg='light blue')\n",
    "    update_window.geometry(\"600x400\")\n",
    "\n",
    "    # Create the labels, text boxes, and button for updating the toxin data\n",
    "    old_health_effects_label =Label(update_window, text=\"Old Health Effects:\", bg='light blue')\n",
    "    old_health_effects_label.pack()\n",
    "\n",
    "    old_health_effects_entry = Entry(update_window)\n",
    "    old_health_effects_entry.pack()\n",
    "\n",
    "    new_health_effects_label = Label(update_window, text=\"New Health Effects:\", bg='light blue')\n",
    "    new_health_effects_label.pack()\n",
    "\n",
    "    new_health_effects_entry = Entry(update_window)\n",
    "    new_health_effects_entry.pack()\n",
    "\n",
    "    update_button = Button(update_window, text=\"Update\", command=lambda: update_table())\n",
    "    update_button.pack()\n",
    "\n",
    "    def update_table():\n",
    "        # Get the old and new health effects values and update the collection\n",
    "        old_health_effects = old_health_effects_entry.get()\n",
    "        new_health_effects = new_health_effects_entry.get()\n",
    "        df = update_toxin_data_collection(old_health_effects, new_health_effects)\n",
    "\n",
    "        # Clear the previous contents of the table frame\n",
    "        for widget in table_frame.winfo_children():\n",
    "            widget.destroy()\n",
    "\n",
    "        # Create the table using pandas and display it in the table frame\n",
    "        table = pd.DataFrame(df)\n",
    "        table_widget = Text(table_frame, bg='white', relief=\"groove\", bd=2)\n",
    "        table_widget.insert(\"end\", table.to_string(index=False))\n",
    "        table_widget.pack()\n",
    "\n",
    "    close_button = Button(update_window, text=\"Close\", command=update_window.destroy)\n",
    "    close_button.pack()\n",
    "\n",
    "    # Create the label to show the updated documents\n",
    "    results_label = Label(update_window, text=\"Results:\", bg='light blue')\n",
    "    results_label.pack()\n",
    "\n",
    "    # Create the table frame to display the updated documents\n",
    "    table_frame = Frame(update_window, bg='white', relief=\"groove\", bd=2)\n",
    "    table_frame.pack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_synonms():\n",
    "    # Create the GUI\n",
    "    modify_synonms_window = Toplevel()\n",
    "    modify_synonms_window.title(\"Update synonms\")\n",
    "    modify_synonms_window.configure(bg='light blue') # Set the background color to baby blue\n",
    "\n",
    "    # Create the labels and text boxes for the 'old synonyms'and 'new synonyms' fields\n",
    "    old_synonyms_label = Label(modify_synonms_window, text=\"Old Synonyms:\", bg='light blue')\n",
    "    old_synonyms_label.grid(row=0, column=0)\n",
    "    old_synonyms_entry = Entry(modify_synonms_window, width=30)\n",
    "    old_synonyms_entry.grid(row=0, column=1)\n",
    "\n",
    "    new_synonyms_label = Label(modify_synonms_window, text=\"New Synonyms:\", bg='light blue')\n",
    "    new_synonyms_label.grid(row=1, column=0)\n",
    "    new_synonyms_entry = Entry(modify_synonms_window, width=30)\n",
    "    new_synonyms_entry.grid(row=1, column=1)\n",
    "\n",
    "    # Create the update button\n",
    "    update_button = Button(modify_synonms_window, text=\"Update\", command=lambda: update_toxin_target_collection(old_synonyms_entry.get(), new_synonyms_entry.get()))\n",
    "    update_button.grid(row=2, column=0, columnspan=2)\n",
    "    close_button = Button(modify_synonms_window, text=\"Close\", command=modify_synonms_window.destroy)\n",
    "    close_button.grid(row=3, column=0, columnspan=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### update date "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "from pymongo import MongoClient\n",
    "import pandas as pd\n",
    "\n",
    "def update_toxin_data(var1: str, var2: str) -> pd.DataFrame:\n",
    "    # Connect to the MongoDB server and select the database and collection\n",
    "    client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "    db = client[\"toxins\"]\n",
    "    all_toxin_data_collection = db[\"AllToxinData\"]\n",
    "\n",
    "    # Find the document with the given _id and retrieve its current updated_at value\n",
    "    document = all_toxin_data_collection.find_one({\"_id\": var1})\n",
    "\n",
    "    if document is not None:\n",
    "        old_updated_at = document.get(\"updated_at\")\n",
    "\n",
    "        # Update the document with the new updated_at value\n",
    "        newvalues = {\"$set\": {\"updated_at\": var2}}\n",
    "        all_toxin_data_collection.update_one({\"_id\": var1}, newvalues)\n",
    "\n",
    "        # Retrieve the updated document\n",
    "        updated_document = all_toxin_data_collection.find_one({\"_id\": var1})\n",
    "\n",
    "        # Prepare the data to create a DataFrame\n",
    "        data = [\n",
    "            {\"_id\": updated_document.get(\"_id\"), \"old_updated_at\": old_updated_at, \"new_updated_at\": updated_document.get(\"updated_at\")}\n",
    "        ]\n",
    "        df = pd.DataFrame(data)\n",
    "        return df\n",
    "    else:\n",
    "        # Return an empty DataFrame if the document is not found\n",
    "        return pd.DataFrame()\n",
    "\n",
    "def update_toxin_data_gui():\n",
    "    update_date = tk.Toplevel(root)\n",
    "\n",
    "    def update_data():\n",
    "        var1 = id_entry.get()\n",
    "        var2 = new_updated_at_entry.get()\n",
    "        df = update_toxin_data(var1, var2)\n",
    "        result_text.delete(\"1.0\", tk.END)\n",
    "        result_text.insert(tk.END, df.to_string(index=False))\n",
    "\n",
    "    # Create the labels and entry fields\n",
    "    id_label = tk.Label(update_date, text=\"Enter _id:\", bg=\"light blue\")\n",
    "    id_label.grid(row=0, column=0, padx=5, pady=5)\n",
    "    id_entry = tk.Entry(update_date)\n",
    "    id_entry.grid(row=0, column=1, padx=5, pady=5)\n",
    "\n",
    "    new_updated_at_label = tk.Label(update_date, text=\"Enter new updated_at:\", bg=\"light blue\")\n",
    "    new_updated_at_label.grid(row=1, column=0, padx=5, pady=5)\n",
    "    new_updated_at_entry = tk.Entry(update_date)\n",
    "    new_updated_at_entry.grid(row=1, column=1, padx=5, pady=5)\n",
    "\n",
    "    # Create the update button\n",
    "    update_button = tk.Button(update_date, text=\"Update\", command=update_data)\n",
    "    update_button.grid(row=2, column=0, columnspan=2, padx=5, pady=5)\n",
    "\n",
    "    # Create the result label and text box\n",
    "    result_label = tk.Label(update_date, text=\"Result:\", bg=\"light blue\")\n",
    "    result_label.grid(row=3, column=0, padx=5, pady=5)\n",
    "    result_text = tk.Text(update_date, height=5, width=60)\n",
    "    result_text.grid(row=4, column=0, columnspan=2, padx=5, pady=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### update locus value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_toxin_target_collection(var1: str, var2: str) -> pd.DataFrame:\n",
    "    # Connect to the MongoDB server and select the database and collection\n",
    "    client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "    db = client[\"toxins\"]\n",
    "    all_toxin_target_collection = db[\"AllToxinTarget\"]\n",
    "\n",
    "    # Update the documents with the old 'Locus' value to the new value\n",
    "    myquery = {\"Locus\": var1 }\n",
    "    newvalues = {\"$set\": {\"Locus\": var2 } }\n",
    "    all_toxin_target_collection.update_one(myquery, newvalues)\n",
    "\n",
    "    # Find and return the documents with the new 'Locus' value\n",
    "    documents = all_toxin_target_collection.find({\"Locus\":var2})\n",
    "    result = list(documents)\n",
    "    df = pd.DataFrame(result)\n",
    "    return df\n",
    "\n",
    "def update_toxin_target_gui():\n",
    "    # Create the GUI\n",
    "    update_locus = Toplevel()\n",
    "    update_locus.title(\"Update Toxin Target\")\n",
    "    update_locus.configure(bg='light blue') # Set the background color to baby blue    # Create the labels and text boxes for the 'old Locus' and 'new Locus' values\n",
    "    old_locus_label = Label(update_locus, text=\"Old Locus:\", bg='light blue')\n",
    "    old_locus_label.grid(row=0, column=0, padx=5, pady=5)\n",
    "    old_locus_entry = Entry(update_locus)\n",
    "    old_locus_entry.grid(row=0, column=1, padx=5, pady=5)\n",
    "\n",
    "    new_locus_label = Label(update_locus, text=\"New Locus:\", bg='light blue')\n",
    "    new_locus_label.grid(row=1, column=0, padx=5, pady=5)\n",
    "    new_locus_entry = Entry(update_locus)\n",
    "    new_locus_entry.grid(row=1, column=1, padx=5, pady=5)\n",
    "\n",
    "    # Create the button to update the collection\n",
    "    update_button = Button(update_locus, text=\"Update\", bg='white', command=lambda: update_table())\n",
    "    update_button.grid(row=2, column=0, columnspan=2, padx=5, pady=5)\n",
    "\n",
    "    # Create the label to show the updated documents\n",
    "    results_label = Label(update_locus, text=\"Results:\", bg='light blue')\n",
    "    results_label.grid(row=3, column=0, padx=5, pady=5)\n",
    "\n",
    "    # Create the table to display the updated documents\n",
    "    table_frame = Frame(update_locus, bg='white', relief=\"groove\", bd=2)\n",
    "    table_frame.grid(row=4, column=0, columnspan=2, padx=5, pady=5)\n",
    "\n",
    "    def update_table():\n",
    "        \n",
    "        # Get the old and new Locus values and update the collection\n",
    "        old_locus = old_locus_entry.get()\n",
    "        new_locus = new_locus_entry.get()\n",
    "        df = update_toxin_target_collection(old_locus, new_locus)\n",
    "\n",
    "        # Clear the previous contents of the table frame\n",
    "        for widget in table_frame.winfo_children():\n",
    "            widget.destroy()\n",
    "\n",
    "        if not df.empty:\n",
    "            # Select only the \"Locus\" and \"_id\" columns\n",
    "            df_filtered = df[[\"Locus\", \"_id\"]]\n",
    "\n",
    "            # Create the table using pandas and display it in the table frame\n",
    "            table_widget = Text(table_frame, bg='white', relief=\"groove\", bd=2)\n",
    "            table_widget.insert(\"end\", df_filtered.to_string(index=False))\n",
    "            table_widget.pack()\n",
    "        else:\n",
    "            no_results_label = Label(table_frame, text=\"No results found.\", bg='white', relief=\"groove\", bd=2)\n",
    "            no_results_label.pack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_update_target_data():\n",
    "    # Create the top-level window\n",
    "    update_window_target = Toplevel(root)\n",
    "    update_window_target.title(\"Update Target Data\")\n",
    "    update_window_target.configure(bg='steelblue4')\n",
    "\n",
    "    # Set the width and height of the window\n",
    "    window_width = 400\n",
    "    window_height = 300\n",
    "    update_window_target.geometry(f\"{window_width}x{window_height}\")\n",
    "\n",
    "    # Create the buttons\n",
    "    button1 = Button(update_window_target, text=\"Update Synonyms\", command=open_update_screen)\n",
    "    button1.pack(pady=10)\n",
    "\n",
    "    button2 = Button(update_window_target, text=\"Update Locus\", command=update_toxin_target_gui)\n",
    "    button2.pack(pady=10)\n",
    "\n",
    "    # Create the close button\n",
    "    close_button = Button(update_window_target, text=\"Close\", command=update_window_target.destroy)\n",
    "    close_button.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_update_window_entry():\n",
    "    # Create the top-level window\n",
    "    update_window = Toplevel(root)\n",
    "    update_window.title(\"Update Window\")\n",
    "    update_window.configure(bg='steelblue4')\n",
    "\n",
    "    # Set the width and height of the window\n",
    "    window_width = 400\n",
    "    window_height = 300\n",
    "    update_window.geometry(f\"{window_width}x{window_height}\")\n",
    "\n",
    "    # Create the buttons\n",
    "    button1 = Button(update_window, text=\"Open Update Target Data\", command=open_update_target_data)\n",
    "    button1.pack(pady=10)\n",
    "\n",
    "    button2 = Button(update_window, text=\"Open Update Toxin Data\", command=open_update_window)\n",
    "    button2.pack(pady=10)\n",
    "\n",
    "    close_button = Button(update_window, text=\"Close\", command=update_window.destroy)\n",
    "    close_button.pack(pady=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_toxinTarget():\n",
    "    global targetWindow\n",
    "\n",
    "    def insert_data():\n",
    "        # Retrieve the input values from the entry fields\n",
    "        _id = entry_id.get()\n",
    "        name = entry_name.get()\n",
    "        uniprot_name = entry_uniprot.get()\n",
    "        organism = entry_organism.get()\n",
    "        taxonomy_id = entry_taxonomy.get()\n",
    "        gene_name = entry_gene.get()\n",
    "\n",
    "        # Create a dictionary with the input values\n",
    "        data = {\n",
    "            \"_id\": _id,\n",
    "            \"Name\": name,\n",
    "            \"UniProt Name\": uniprot_name,\n",
    "            \"Organism\": organism,\n",
    "            \"NCBI Taxonomy Id\": taxonomy_id,\n",
    "            \"Gene Name\": gene_name\n",
    "        }\n",
    "\n",
    "        # Insert the data into the MongoDB collection\n",
    "        try:\n",
    "            # Update with your MongoDB connection details\n",
    "            db = client['toxins']  # Replace 'your_database' with your database name\n",
    "            collection = db[\"AllToxinTarget\"]  # Replace 'your_collection' with your collection name\n",
    "            collection.insert_one(data)\n",
    "            messagebox.showinfo(\"Success\", \"Data inserted successfully!\")\n",
    "        except Exception as e:\n",
    "            messagebox.showerror(\"Error\", str(e))\n",
    "\n",
    "    def open_update_window():\n",
    "        \n",
    "        update_toxinTarget()\n",
    "\n",
    "    targetWindow = Toplevel(root)\n",
    "    targetWindow.title(\"Insert Data\")\n",
    "    targetWindow.geometry(\"500x400\")\n",
    "    targetWindow.configure(bg=\"lavender\")  # Set background color to lavender\n",
    "\n",
    "    # Create labels and entry fields for each value\n",
    "    label_id = Label(targetWindow, text=\"ID:\")\n",
    "    label_id.pack(anchor=\"center\")\n",
    "    entry_id = Entry(targetWindow)\n",
    "    entry_id.pack(anchor=\"center\")\n",
    "\n",
    "    label_name = Label(targetWindow, text=\"Name:\")\n",
    "    label_name.pack(anchor=\"center\")\n",
    "    entry_name = Entry(targetWindow)\n",
    "    entry_name.pack(anchor=\"center\")\n",
    "\n",
    "    label_uniprot = Label(targetWindow, text=\"UniProt Name:\")\n",
    "    label_uniprot.pack(anchor=\"center\")\n",
    "    entry_uniprot = Entry(targetWindow)\n",
    "    entry_uniprot.pack(anchor=\"center\")\n",
    "\n",
    "    label_organism = Label(targetWindow, text=\"Organism:\")\n",
    "    label_organism.pack(anchor=\"center\")\n",
    "    entry_organism = Entry(targetWindow)\n",
    "    entry_organism.pack(anchor=\"center\")\n",
    "\n",
    "    label_taxonomy = Label(targetWindow, text=\"NCBI Taxonomy Id:\")\n",
    "    label_taxonomy.pack(anchor=\"center\")\n",
    "    entry_taxonomy = Entry(targetWindow)\n",
    "    entry_taxonomy.pack(anchor=\"center\")\n",
    "\n",
    "    label_gene = Label(targetWindow, text=\"Gene Name:\")\n",
    "    label_gene.pack(anchor=\"center\")\n",
    "    entry_gene = Entry(targetWindow)\n",
    "    entry_gene.pack(anchor=\"center\")\n",
    "\n",
    "    # Create the insert button\n",
    "    button_insert = Button(targetWindow, text=\"Insert\", command=insert_data)\n",
    "    button_insert.pack(anchor=\"center\")\n",
    "\n",
    "    # Create the update button\n",
    "    button_update = Button(targetWindow, text=\"Update\", command=open_update_target_data)\n",
    "    button_update.pack(anchor=\"center\")\n",
    "    \n",
    "    find_btn=Button(targetWindow,text=\"find Queries\",command=open_find_queries_window_target)\n",
    "    find_btn.pack(anchor=\"center\")\n",
    "\n",
    "    # Create the close button\n",
    "    button_close = Button(targetWindow, text=\"Close\", command=targetWindow.destroy)\n",
    "    button_close.pack(anchor=\"center\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Analysis Windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def close_window():\n",
    "\n",
    "    graph_window.destroy()\n",
    "    \n",
    "def close_cloud():\n",
    "    cloud_window.destroy()\n",
    "    \n",
    "def close_query():\n",
    "    Querywindow.destroy()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_dataAnalysis():\n",
    "    # Create a new Toplevel window for the new graph\n",
    "    data_AnalysisWindow = Toplevel(root)\n",
    "    data_AnalysisWindow.title(\"Data Analysis Window\")\n",
    "    data_AnalysisWindow.geometry(\"800x600\")\n",
    "    #data_AnalysisWindow.configure(bg=\"blue4\")\n",
    "    background_image = PhotoImage(file=\"graphs.PNG\")\n",
    "    background_label = Label(data_AnalysisWindow, image=background_image)\n",
    "    background_label.image = background_image  # Retain a reference to the image\n",
    "    background_label.pack()\n",
    "    \n",
    "    #################Show scatter plot########################\n",
    "    def show_graph():\n",
    "        global graph_window\n",
    "        # Create a new Toplevel window\n",
    "        graph_window =Toplevel(root)\n",
    "        graph_window.configure(bg=\"gold\")\n",
    "        graph_window.title(\"Graph\")\n",
    "        # Create a Figure object\n",
    "        fig = Figure(figsize=(6, 4), dpi=100)\n",
    "        # Add a subplot to the figure\n",
    "        ax = fig.add_subplot(111)\n",
    "        df = pd.read_csv('AllToxinTarget.csv')\n",
    "        col1 = 'Genbank Protein ID'  # the first column containing the data for the x-axis\n",
    "        col2 = 'Number of Residues'  # the second column containing the data for the y-axis\n",
    "        # select the first 10 rows of the columns\n",
    "        data1 = df.head(20)[col1]\n",
    "        data2 = df.head(20)[col2]\n",
    "\n",
    "        # Plot your data (example: a simple line plot)\n",
    "\n",
    "        ax.scatter(data1, data2)\n",
    "        ax.set_xlabel('Genbank Protein ID')\n",
    "        ax.set_ylabel('Number of Residues')\n",
    "        ax.set_title(\"Genbank ID and Number of Residues\")\n",
    "\n",
    "        # Create a canvas for displaying the figure\n",
    "        canvas = FigureCanvasTkAgg(fig, master=graph_window)\n",
    "        canvas.draw()\n",
    "        canvas.get_tk_widget().pack()\n",
    "\n",
    "        close_button =Button(graph_window,text=\"Close\",width=10, command=close_window)\n",
    "        close_button.pack()\n",
    "        \n",
    "    ########################draw theortical phi graph #########################\n",
    "    def draw_theoretical_phi():\n",
    "        \n",
    "        # Create a new tkinter window\n",
    "        global draw_theoretical_phi_window\n",
    "\n",
    "        draw_theoretical_phi_window = Toplevel(root)\n",
    "        draw_theoretical_phi_window.title(\"Theoretical Pi by NCBI Taxonomy Id\")\n",
    "\n",
    "        # Load the CSV file into a pandas dataframe\n",
    "        df = pd.read_csv('AllToxinTarget.csv')\n",
    "        df = df.head(40)\n",
    "\n",
    "        # Convert NCBI Taxonomy Id column to string and Theoretical Pi column to float\n",
    "        df['NCBI Taxonomy Id'] = df['NCBI Taxonomy Id'].astype(str)\n",
    "        df['Theoretical Pi'] = pd.to_numeric(df['Theoretical Pi'], errors='coerce')\n",
    "\n",
    "        # Create a Figure object\n",
    "        fig = plt.Figure(figsize=(6, 4), dpi=100)\n",
    "\n",
    "        # Create a subplot in the figure\n",
    "        ax = fig.add_subplot(111)\n",
    "\n",
    "        # Create a hexbin plot using matplotlib\n",
    "        ax.hexbin(df['NCBI Taxonomy Id'], df['Theoretical Pi'], gridsize=20)\n",
    "\n",
    "        # Add title and labels\n",
    "        ax.set_title('Theoretical Pi by NCBI Taxonomy Id (First 10 values)')\n",
    "        ax.set_xlabel('NCBI Taxonomy Id')\n",
    "        ax.set_ylabel('Theoretical Pi')\n",
    "\n",
    "        # Create a canvas for displaying the plot\n",
    "        canvas = FigureCanvasTkAgg(fig, master=draw_theoretical_phi_window)\n",
    "        canvas.draw()\n",
    "        canvas.get_tk_widget().pack()\n",
    "\n",
    "        # Create a function to close the window\n",
    "        def close_window():\n",
    "            draw_theoretical_phi_window.destroy()\n",
    "\n",
    "        # Create a close button\n",
    "        close_button = Button(draw_theoretical_phi_window, text=\"Close\", command=close_window)\n",
    "        close_button.pack()\n",
    "        \n",
    "    ########################## barplott function####################################\n",
    "    def show_barplot():\n",
    "        \n",
    "        global barwindow\n",
    "        # Create a new tkinter window\n",
    "        barwindow = Toplevel(root)\n",
    "        root.title(\"Toxin State Countplot\")\n",
    "\n",
    "        # Load the CSV file into a pandas dataframe\n",
    "        df = pd.read_csv('AllToxinData.csv')\n",
    "        df = df.head(10)\n",
    "\n",
    "        # Create a Figure object\n",
    "        fig = plt.Figure(figsize=(6, 4), dpi=100)\n",
    "\n",
    "        # Create a subplot in the figure\n",
    "        ax = fig.add_subplot(111)\n",
    "\n",
    "        # Create a countplot using seaborn with custom color palette\n",
    "        sns.countplot(x='common_name', hue='state', data=df, palette=['purple', 'yellow', 'blue'], ax=ax)\n",
    "\n",
    "        # Rotate the x-axis labels to avoid overlap\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "\n",
    "        # Add title and labels\n",
    "        ax.set_title('Toxin State by Name')\n",
    "        ax.set_xlabel('Toxin Name')\n",
    "        ax.set_ylabel('Count')\n",
    "\n",
    "        # Create a canvas for displaying the plot\n",
    "        canvas = FigureCanvasTkAgg(fig, master=barwindow)\n",
    "        canvas.draw()\n",
    "        canvas.get_tk_widget().pack()\n",
    "\n",
    "        # Create a function to close the window\n",
    "        def close_window():\n",
    "            barwindow.destroy()\n",
    "\n",
    "        # Create a close button\n",
    "        close_button =Button(barwindow, text=\"Close\", command=close_window)\n",
    "        close_button.pack()\n",
    "\n",
    "############################################show box plot#####################################\n",
    "    def show_box_plots():\n",
    "        global boxplot_window\n",
    "        boxplot_window=Toplevel(root)\n",
    "        # Load the data from the CSV file\n",
    "        df = pd.read_csv('AllToxinData.csv')\n",
    "\n",
    "        # Specify the columns to use for the box plots\n",
    "        col1 = 'weight'\n",
    "        col2 = 'moldb_average_mass'\n",
    "        col3 = 'moldb_mono_mass'\n",
    "\n",
    "        # Select the first 10 rows of the columns and parse the data as floats\n",
    "        data1 = df.head(10)[col1].astype(float)\n",
    "        data2 = df.head(10)[col2].astype(float)\n",
    "        data3 = df.head(10)[col3].astype(float)\n",
    "\n",
    "        # Create the subplots for the box plots\n",
    "        fig, axs = plt.subplots(1, 3, figsize=(10, 5))\n",
    "\n",
    "        # Create a box plot for each column using matplotlib\n",
    "        axs[0].boxplot(data1)\n",
    "        axs[1].boxplot(data2)\n",
    "        axs[2].boxplot(data3)\n",
    "\n",
    "        # Add titles and labels for each box plot\n",
    "        axs[0].set_title('moldb_average_mass')\n",
    "        axs[1].set_title('moldb_mono_mass')\n",
    "        axs[2].set_title('weight')\n",
    "        axs[0].set_xlabel('X-axis Label')\n",
    "        axs[1].set_xlabel('X-axis Label')\n",
    "        axs[2].set_xlabel('X-axis Label')\n",
    "        axs[0].set_ylabel('Y-axis Label')\n",
    "        axs[1].set_ylabel('Y-axis Label')\n",
    "        axs[2].set_ylabel('Y-axis Label')\n",
    "\n",
    "        # Adjust the layout and spacing of the subplots\n",
    "        fig.tight_layout(pad=3.0)\n",
    "\n",
    "        canvas = FigureCanvasTkAgg(fig, master=boxplot_window)\n",
    "        canvas.draw()\n",
    "        canvas.get_tk_widget().pack()\n",
    "        def close_boxwindow():\n",
    "            boxplot_window.destroy()\n",
    "\n",
    "        # Create a Button to close the window\n",
    "        close_button = Button(boxplot_window, text=\"Close\", command=close_boxwindow)\n",
    "        close_button.pack()\n",
    "        \n",
    "    ##############################draw stack plot##################################################33\n",
    "    def draw_area_plot():\n",
    "        # Create a new tkinter window\n",
    "        global areaplot_window\n",
    "        areaplot_window = Toplevel(root)\n",
    "        areaplot_window.title(\"Area Plot\")\n",
    "\n",
    "        # Load the CSV file into a pandas dataframe\n",
    "        df = pd.read_csv('AllToxinData.csv')\n",
    "        df = df.head(10)\n",
    "\n",
    "        # Create a Figure object\n",
    "        fig = plt.Figure(figsize=(6, 4), dpi=100)\n",
    "\n",
    "        # Create a subplot in the figure\n",
    "        ax = fig.add_subplot(111)\n",
    "\n",
    "        # Convert toxicity and name columns to string\n",
    "        df['chemical_formula'] = df['chemical_formula'].astype(str)\n",
    "        df['common_name'] = df['common_name'].astype(str)\n",
    "\n",
    "        # Create an area plot using matplotlib\n",
    "        ax.stackplot(df['common_name'], df['chemical_formula'], alpha=0.7)\n",
    "\n",
    "        # Add title and labels\n",
    "        ax.set_title('Chemical Formula')\n",
    "        ax.set_xlabel('Name')\n",
    "        ax.set_ylabel('Toxicity')\n",
    "\n",
    "        # Create a canvas for displaying the plot\n",
    "        canvas = FigureCanvasTkAgg(fig, master=areaplot_window)\n",
    "        canvas.draw()\n",
    "        canvas.get_tk_widget().pack()\n",
    "\n",
    "        # Create a function to close the window\n",
    "        def close_window():\n",
    "            areaplot_window.destroy()\n",
    "\n",
    "        # Create a close button\n",
    "        close_button =Button(areaplot_window, text=\"Close\", command=close_window)\n",
    "        close_button.pack()\n",
    "        #########################################word cloud button######################################3\n",
    "    def show_wordcloud():\n",
    "        global cloud_window\n",
    "        # Generate the word cloud\n",
    "        fields = ['Toxin Name']\n",
    "        mechanism= pd.read_csv('Relations_table1.csv', usecols=fields)\n",
    "        text2=mechanism[\"Toxin Name\"]\n",
    "        mechanism=text2[2:1500]\n",
    "        text3 = ' '.join(text2)\n",
    "\n",
    "        #wc=WordCloud(max_words=50,background_color=\"white\").generate(text3)  # Replace with your text data\n",
    "        wordcloud = WordCloud(background_color=\"white\").generate(text3)\n",
    "\n",
    "        # Create a Toplevel window\n",
    "        cloud_window = Toplevel(root)\n",
    "        cloud_window.title(\"Word Cloud\")\n",
    "\n",
    "        # Convert the word cloud image to a Tkinter-compatible format\n",
    "        image = wordcloud.to_image()\n",
    "\n",
    "        # Display the word cloud image in a Tkinter Label widget\n",
    "        label =Label(cloud_window)\n",
    "        label.pack()\n",
    "\n",
    "        # Convert the image to PhotoImage format and set it as the Label's image\n",
    "        photo = ImageTk.PhotoImage(image)\n",
    "        label.configure(image=photo)\n",
    "        label.image = photo\n",
    "        close_button =Button(cloud_window, text=\"Close\",width=10, command=close_cloud)\n",
    "        close_button.pack()\n",
    "\n",
    "\n",
    "    # Create a button to show the word cloud\n",
    "    word_cloud_btn = Button(data_AnalysisWindow, text=\"Show Word Cloud\",bg=\"gold\" ,command=show_wordcloud)\n",
    "    word_cloud_btn.pack()\n",
    "    \n",
    "    #create a button  to show box plot\n",
    "    boxplot_btn = Button(data_AnalysisWindow, text=\"Show box plot\",bg=\"gold\" ,command=show_box_plots)\n",
    "    boxplot_btn.pack()\n",
    "    \n",
    "    #create a button to show draw_theoretical_phi graph\n",
    "    \n",
    "    draw_theoretical_phi_btn=Button(data_AnalysisWindow, text=\"Show theoretical_phi graph\",bg=\"gold\" ,command=draw_theoretical_phi)\n",
    "    draw_theoretical_phi_btn.pack()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # Create a button to show the scatter plot\n",
    "    scatter_button = Button(data_AnalysisWindow, text=\"Show Scatter Plot\",bg=\"gold\", command=show_graph)\n",
    "    scatter_button.pack()\n",
    "    #create button to show barplot##################\n",
    "    show_barplot_btn=Button(data_AnalysisWindow, text=\"Show bar Plot\",bg=\"gold\", command=show_barplot)\n",
    "    show_barplot_btn.pack()\n",
    "    \n",
    "    #create a button to show a stack plot\n",
    "    draw_area_plot_btn=Button(data_AnalysisWindow, text=\"Show area plot\",bg=\"gold\", command=draw_area_plot)\n",
    "    draw_area_plot_btn.pack()\n",
    "    \n",
    "\n",
    "\n",
    "    # Calculate the position for the button\n",
    "    button_width = 200\n",
    "    button_height = 50\n",
    "    button_x = (800 - button_width) // 2\n",
    "    button_y = (600 - button_height) // 2\n",
    "\n",
    "    # Create the button\n",
    "    scatter_button.place(x=button_x, y=button_y, width=button_width, height=button_height)\n",
    "    word_cloud_btn.place(x=button_x, y=button_y+50, width=button_width, height=button_height)\n",
    "    boxplot_btn.place(x=button_x, y=button_y+100, width=button_width, height=button_height)\n",
    "    show_barplot_btn.place(x=button_x, y=button_y+150, width=button_width, height=button_height)\n",
    "    draw_theoretical_phi_btn.place(x=button_x, y=button_y+200, width=button_width, height=button_height)\n",
    "    draw_area_plot_btn.place(x=button_x, y=button_y-50, width=button_width, height=button_height)\n",
    "        \n",
    "\n",
    "    # Handle the window close event\n",
    "    def on_close():\n",
    "        if messagebox.askokcancel(\"Close\", \"Do you want to close the Data Analysis window?\"):\n",
    "            data_AnalysisWindow.destroy()\n",
    "\n",
    "    data_AnalysisWindow.protocol(\"WM_DELETE_WINDOW\", on_close)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "from pymongo import MongoClient\n",
    "from pymongo.cursor import Cursor\n",
    "\n",
    "# connect to MongoDB\n",
    "client = MongoClient(\"mongodb://localhost:27017/\")\n",
    "db = client['toxins']\n",
    "collection = db['mycollection']\n",
    "\n",
    "def open_code_executor_window():\n",
    "    code_executor_window = tk.Toplevel(root)\n",
    "    code_executor_window.title(\"Code Executor\")\n",
    "\n",
    "    code_label = tk.Label(code_executor_window, text='Enter code:')\n",
    "    code_label.pack()\n",
    "\n",
    "    code_entry = tk.Entry(code_executor_window, width=50)\n",
    "    code_entry.pack()\n",
    "\n",
    "    execute_button = tk.Button(code_executor_window, text='Execute', command=lambda: execute_code(code_entry, result_text))\n",
    "    execute_button.pack()\n",
    "\n",
    "    result_label = tk.Label(code_executor_window, text='Result:')\n",
    "    result_label.pack()\n",
    "\n",
    "    result_frame = tk.Frame(code_executor_window, bg='gray')\n",
    "    result_frame.pack(fill=tk.BOTH, expand=True)\n",
    "\n",
    "    result_scrollbar = tk.Scrollbar(result_frame)\n",
    "    result_scrollbar.pack(side=tk.RIGHT, fill=tk.Y)\n",
    "\n",
    "    result_text = tk.Text(result_frame, bg='lavender', fg='black', state=tk.DISABLED,\n",
    "                          yscrollcommand=result_scrollbar.set)\n",
    "    result_text.pack(fill=tk.BOTH, expand=True)\n",
    "\n",
    "    result_scrollbar.config(command=result_text.yview)\n",
    "\n",
    "def execute_code(code_entry, result_text):\n",
    "    code = code_entry.get()\n",
    "    try:\n",
    "        # check if code contains a query to the database\n",
    "        if 'db.' in code:\n",
    "            result = eval(code, {'db': db, 'collection': collection})\n",
    "        else:\n",
    "            result = eval(code)\n",
    "\n",
    "        result_text.config(state=tk.NORMAL)\n",
    "        result_text.delete('1.0', tk.END)\n",
    "\n",
    "        # check if result is a cursor object from MongoDB\n",
    "        if isinstance(result, Cursor):\n",
    "            for document in result:\n",
    "                result_text.insert(tk.END, str(document) + '\\n')\n",
    "        else:\n",
    "            result_text.insert(tk.END, str(result))\n",
    "\n",
    "        result_text.config(state=tk.DISABLED)\n",
    "    except Exception as e:\n",
    "        result_text.config(state=tk.NORMAL)\n",
    "        result_text.delete('1.0', tk.END)\n",
    "        result_text.insert(tk.END, str(e))\n",
    "        result_text.config(state=tk.DISABLED)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from tkinter import Tk, Label, Button\n",
    "from PIL import Image, ImageTk\n",
    "\n",
    "root = Tk()\n",
    "\n",
    "# Create a photoimage object of the image in the path\n",
    "image1 = Image.open(\"homeScreen.PNG\")\n",
    "test = ImageTk.PhotoImage(image1)\n",
    "\n",
    "root.title(\"Toxins database\")\n",
    "\n",
    "label1 = Label(image=test)\n",
    "label1.image = test\n",
    "label1.place(x=0, y=0)\n",
    "\n",
    "# Define functions for button commands (e.g., open_dataAnalysis, insertion_screen)\n",
    "\n",
    "Button(root, width=25, height=2, text=\"Data Analysis\", background=\"SpringGreen3\", foreground=\"blue4\",\n",
    "       font=('Helvetica 13 bold'), command=open_dataAnalysis).place(x=590, y=500)\n",
    "Button(root, width=25, height=2, text=\"Execute your  Query \", background=\"SpringGreen3\", foreground=\"blue4\",\n",
    "       font=('Helvetica 13 bold'), command=open_code_executor_window).place(x=300, y=500)\n",
    "\n",
    "Button(root, width=25, height=2, text=\"Data query\", background=\"SpringGreen3\", foreground=\"blue4\",\n",
    "       font=('Helvetica 13 bold'), command=query_window).place(x=10, y=500)\n",
    "\n",
    "root.geometry(\"850x600\")\n",
    "root.mainloop()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### update date all toxin data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"Molecular Weight\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
